{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f9f70a4d-d0a8-4e42-9fcc-7874e7868834",
   "metadata": {},
   "source": [
    "Hui Yu's answers to these three questions.\n",
    "\n",
    "What was your biggest challenge in this project? (regarding writing code and not only)\n",
    "\n",
    "The biggest challenge is to understand K-means. It took me quite a while to read through the file and understand the concept. I was quite confused at begining why the accuracy number is different each time I ran. I will remember to set up random_state to make my results reproducible. \n",
    "\n",
    "What did you learn while working on this project? (regarding writing code and not only)\n",
    "\n",
    "I noticed that the KMeans method is acturally very common used. In my own single cell analysis, the k-mean is widely used to cluster cells and generate UMAP plot.I was not fully aware of it until I read through the kmean document from this project. \n",
    "When seperating train and test datasets, I was thinking to manually seperate half dataset for train dataset and half for test dataset, however, my group member used  sklearn.model_selection train_test_split function. Very smart. \n",
    "\n",
    "\n",
    "If you had more time on the project what other question(s) would you like to answer? (at least one question is required)\n",
    "\n",
    "I would like to compare the three accuracy numbers and investigate why I got the highest accuracy score at the train dataset and why the accuracy scores have this trend in the three tests? What can be possibly done to improve the accuracy score (from both sample collections and fitting to approprate model point of views)?\n",
    "\n",
    "----------------------------------------------------------------------------------------------------------------\n",
    "\n",
    "Keagan Moo's answers to the three questions:\n",
    "\n",
    "What was your biggest challenge in this project? (regarding writing code and not only)\n",
    "\n",
    "The biggest coding challenge was wrapping my head around the structure of the GEOParse object and making sure I was getting the correct pieces out of it. \n",
    "\n",
    "What did you learn while working on this project? (regarding writing code and not only)\n",
    "\n",
    "I learned that sklearn designates its own internal state for reproducibility rather than depending on the environment for the seed. That had not been my experience with similar tools in R or Python so it was good to know. \n",
    "\n",
    "If you had more time on the project what other question(s) would you like to answer? (at least one question is required)\n",
    "\n",
    "I'd like to ask if a consensus model of some kind could be generated by training the model with thirds of the data and combining the results in some meaningful way. Then of course I'd like to know how that stacks up against other methods.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48f5cff8-f8b9-4c8e-a64c-3339dffbfdcd",
   "metadata": {},
   "source": [
    "Maya Bose question responses\n",
    "1. My biggest challenge for the project was understanding how the part of the project I was working on fit into the bigger picture of the assignment. I had never worked on coding projects in groups before, and so learning how to understand what other people's code was doing and how to work with it was very new to me. \n",
    "\n",
    "2. I learned that scikit learn has a function for dividing a dataset into a testing set and a training set called test_train_split. \n",
    "\n",
    "3. If I had more time to work of the project, I would like to explore what parameters may optimize the accuracy of the KMeans clustering, or explore if there are other classification methods that may fit the situation better. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "a73164a8-ba3c-478e-8ff4-56af31ddba6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import GEOparse\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.metrics import accuracy_score\n",
    "import numpy as np\n",
    "from sklearn.model_selection import train_test_split\n",
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "random_seed = 3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd0de012-d3e6-44ed-a48c-9736a28260e6",
   "metadata": {},
   "source": [
    "# 1 - Use the GEOparse module to load the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "fa6730d2-7ade-44d8-9001-9036aaecce3b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mbose/miniconda3/lib/python3.9/site-packages/GEOparse/GEOparse.py:545: DtypeWarning: Columns (2) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  table_data = parse_table_data(gpl_soft)\n"
     ]
    }
   ],
   "source": [
    "# Use the lovely GEO API to get the default version of the data at that address, in this case it is the\n",
    "# soft zipped file which is what we want\n",
    "rawDataGSE10245 = GEOparse.get_GEO(geo=\"GSE10245\", destdir=\"./\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "753d8304-9824-4fea-8eb7-1452e557464b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'GEOparse.GEOTypes.GSE'>\n",
      "Non-small cell lung cancer (NSCLC) can be classified into the major subtypes adenocarcinoma (AC) and squamous cell carcinoma (SCC) subtypes. Although explicit molecular, histological and clinical characteristics have been reported for both subtypes, no specific therapy exists so far. However, the characterization of suitable molecular targets holds great promises to develop novel therapies in NSCLC. In the present study, global gene expression profiling of 58 human high grade NSCLC specimens revealed large transcriptomic differences between AC and SCC subtypes: More than 1.700 genes were found to be differentially expressed.\n",
      "\n",
      "Sample GSM258551\n",
      "('title', ['NSCLC_AC_10'])\n",
      "('geo_accession', ['GSM258551'])\n",
      "('status', ['Public on Oct 01 2009'])\n",
      "('submission_date', ['Jan 23 2008'])\n",
      "('last_update_date', ['Aug 28 2018'])\n",
      "('type', ['RNA'])\n",
      "('channel_count', ['1'])\n",
      "('source_name_ch1', ['human non-small cell lung cancer tumor tissue'])\n",
      "('organism_ch1', ['Homo sapiens'])\n",
      "('taxid_ch1', ['9606'])\n",
      "('characteristics_ch1', ['disease state: adenocarcinoma'])\n",
      "('treatment_protocol_ch1', ['tumor tissue samples from surgically treated patients were collected intraoperatively, snap-frozen and stored at -80 C. All samples underwent a careful analysis of the histopathology, estimation of tumor and stromal content and subsequent macro-dissection of the tumor regions before enrollment into the microarray experiments.'])\n",
      "('molecule_ch1', ['total RNA'])\n",
      "('extract_protocol_ch1', ['total RNA was extracted using the RNeasy Kit (Qiagen, Hilden, Germany) according to the manufacturersÆ instructions.'])\n",
      "('label_ch1', ['biotin'])\n",
      "('label_protocol_ch1', ['Biotinylated cRNA were prepared according to the standard Affymetrix protocol from 2 ug total RNA (Affymetrix ROC technical manual 701021 Rev.5).'])\n",
      "('hyb_protocol', ['Following fragmentation, 10 ug of cRNA were hybridized for 16 hr at 45C on Affymetrix GeneChip« U133 Plus 2.0 in the Affymetrix GeneChip« Hybridization Oven 640. GeneChips were washed and stained in the Affymetrix GeneChip« Fluidics Station 450.'])\n",
      "('scan_protocol', ['Hybridized microarrays were scanned with the Affymetrix GeneChip« Scanner 3000.'])\n",
      "('description', ['Gene expression data from human non-small cell lung cancer tumor tissue 10.'])\n",
      "('data_processing', ['Bioconductor [Gentleman et al. Genome Biology 2004] and R language [Ihaka and Gentleman. Journal of Computational and Graphical Statistics 1996] were used for pre-processing of the 58 hybridizations. The microarray data was normalized by the gcRMA method [Wu et al. Journal of the American Statistical Association 2004].'])\n",
      "('platform_id', ['GPL570'])\n",
      "('contact_name', ['Ruprecht,,Kuner'])\n",
      "('contact_laboratory', ['Unit Cancer Genome Research'])\n",
      "('contact_department', ['Molecular Genetics'])\n",
      "('contact_institute', ['German Cancer Research Center and National Center of Tumor Diseases'])\n",
      "('contact_address', ['Im Neuenheimer Feld 460'])\n",
      "('contact_city', ['Heidelberg'])\n",
      "('contact_zip/postal_code', ['69120'])\n",
      "('contact_country', ['Germany'])\n",
      "('supplementary_file', ['ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM258nnn/GSM258551/suppl/GSM258551.CEL.gz'])\n",
      "('relation', ['Reanalyzed by: GSE60486', 'Reanalyzed by: GSE119087'])\n",
      "('series_id', ['GSE10245'])\n",
      "('data_row_count', ['54675'])\n",
      "\n",
      "Platform GPL570\n",
      "('title', ['[HG-U133_Plus_2] Affymetrix Human Genome U133 Plus 2.0 Array'])\n",
      "('geo_accession', ['GPL570'])\n",
      "('status', ['Public on Nov 07 2003'])\n",
      "('submission_date', ['Nov 07 2003'])\n",
      "('last_update_date', ['Dec 14 2020'])\n",
      "('technology', ['in situ oligonucleotide'])\n",
      "('distribution', ['commercial'])\n",
      "('organism', ['Homo sapiens'])\n",
      "('taxid', ['9606'])\n",
      "('manufacturer', ['Affymetrix'])\n",
      "('manufacture_protocol', [\"see manufacturer's web site\", '', 'Complete coverage of the Human Genome U133 Set plus 6,500 additional genes for analysis of over 47,000 transcripts', 'All probe sets represented on the GeneChip Human Genome U133 Set are identically replicated on the GeneChip Human Genome U133 Plus 2.0 Array. The sequences from which these probe sets were derived were selected from GenBank®, dbEST, and RefSeq. The sequence clusters were created from the UniGene database (Build 133, April 20, 2001) and then refined by analysis and comparison with a number of other publicly available databases, including the Washington University EST trace repository and the University of California, Santa Cruz Golden-Path human genome database (April 2001 release).', 'In addition, there are 9,921 new probe sets representing approximately 6,500 new genes. These gene sequences were selected from GenBank, dbEST, and RefSeq. Sequence clusters were created from the UniGene database (Build 159, January 25, 2003) and refined by analysis and comparison with a number of other publicly available databases, including the Washington University EST trace repository and the NCBI human genome assembly (Build 31).', ''])\n",
      "('description', ['Affymetrix submissions are typically submitted to GEO using the GEOarchive method described at http://www.ncbi.nlm.nih.gov/projects/geo/info/geo_affy.html', '', 'June 03, 2009: annotation table updated with netaffx build 28', 'June 06, 2012: annotation table updated with netaffx build 32', 'June 23, 2016: annotation table updated with netaffx build 35'])\n",
      "('web_link', ['http://www.affymetrix.com/support/technical/byproduct.affx?product=hg-u133-plus', 'http://www.affymetrix.com/analysis/index.affx'])\n",
      "('contact_name', [',,Affymetrix, Inc.'])\n",
      "('contact_email', ['geo@ncbi.nlm.nih.gov, support@affymetrix.com'])\n",
      "('contact_phone', ['888-362-2447'])\n",
      "('contact_institute', ['Affymetrix, Inc.'])\n",
      "('contact_address', [''])\n",
      "('contact_city', ['Santa Clara'])\n",
      "('contact_state', ['CA'])\n",
      "('contact_zip/postal_code', ['95051'])\n",
      "('contact_country', ['USA'])\n",
      "('contact_web_link', ['http://www.affymetrix.com/index.affx'])\n",
      "('relation', ['Alternative to: GPL4454 (Alternative CDF)', 'Alternative to: GPL4866 (Alternative CDF)', 'Alternative to: GPL5760 (Alternative CDF)', 'Alternative to: GPL6671 (Alternative CDF)', 'Alternative to: GPL6791 (Alternative CDF)', 'Alternative to: GPL6879 (Alternative CDF)', 'Alternative to: GPL7567 (Alternative CDF)', 'Alternative to: GPL8019 (Alternative CDF)', 'Alternative to: GPL8542 (Alternative CDF)', 'Alternative to: GPL8715 (Alternative CDF)', 'Alternative to: GPL8712 (Alternative CDF)', 'Alternative to: GPL9102 (Probe Level Version)', 'Alternative to: GPL9099 (Alternative CDF)', 'Alternative to: GPL9101 (Alternative CDF)', 'Alternative to: GPL9324 (Alternative CDF)', 'Alternative to: GPL9486 (Alternative CDF)', 'Alternative to: GPL9987 (Alternative CDF)', 'Alternative to: GPL10175 (Alternative CDF)', 'Alternative to: GPL10335 (Alternative CDF)', 'Alternative to: GPL10371 (Alternative CDF)', 'Alternative to: GPL10526 (Alternative CDF)', 'Alternative to: GPL10881 (Alternative CDF)', 'Alternative to: GPL10925 (Alternative CDF)', 'Alternative to: GPL11084 (Alternative CDF)', 'Alternative to: GPL11433 (Alternative CDF)', 'Alternative to: GPL11670 (Alternative CDF)', 'Alternative to: GPL13232 (Alternative CDF [HGU133Plus2_Hs_REFSEQ, Brainarray version 13.0.0])', 'Alternative to: GPL13303 (Alternative CDF.)', 'Alternative to: GPL13916 ([Brainarray ENSG Version 14.1])', 'Alternative to: GPL14837 (Alternative CDF)', 'Alternative to: GPL14877 ([Alternative CDF])', 'Alternative to: GPL15394 (Alternative CDF [HGU133Plus2_Hs_REFSEQ, Brainarray version 14.0])', 'Alternative to: GPL6732 (Alternative CDF [HGU133Plus2_Hs_REFSEQ, Brainarray version 7])', 'Alternative to: GPL15445 (Alternative CDF)', 'Alternative to: GPL16066 (Alternative CDF)', 'Alternative to: GPL16100 (Alternative CDF [GATExplorer_Ensembl v57])', 'Alternative to: GPL16268 (Alternative CDF [Brainarray ENSG Version 16])', 'Alternative to: GPL16273 (Alternative CDF [Brainarray ENSG Version 15])', 'Alternative to: GPL16311 (Alternative CDF [HGU133Plus2_Hs_ENTREZG Brainarray version 14.0.0])', 'Alternative to: GPL16356 (Alternative CDF [Brainarray HGU133Plus2_Hs_ENTREZG 15.0.0])', 'Alternative to: GPL16372 (Alternative CDF [Brainarray HGU133Plus2_Hs_UG v.13])', 'Alternative to: GPL17392 (Alternative CDF [HGU133Plus2_Hs_ENSG, Brainarray 17.1])', 'Alternative to: GPL17394 (Alternative CDF [HGU133Plus2_Hs_ENSG, Brainarray 13.0])', 'Alternative to: GPL17810 (Alternative CDF [HGU133Plus2_Hs_ENTREZG, Brainarray 16])', 'Alternative to: GPL17811 (Alternative CDF [HGU133Plus2_Hs_ENTREZG, Brainarray 17])', 'Alternative to: GPL17996 (Alternative CDF:HGU133Plus2_Hs_ENTREZG.cdf version 15.1.0)', 'Alternative to: GPL18121 (custom CDF)', 'Alternative to: GPL18756 (Alternative CDF [CDF: HGU133Plus2_Hs_ENTREZG_14.1.0])', 'Alternative to: GPL18850 (Alternative CDF)', 'Alternative to: GPL18478 (Alternative CDF [Hs_ENTREZG_V13])', 'Alternative to: GPL19109 (Alternative CDF [Brainarray HGU133Plus2_Hs_ENTREZG_v18])', 'Alternative to: GPL19883 (Alternative CDF [Broad GenePattern])', 'Alternative to: GPL19918 (Gene symbol version, 10K)', 'Alternative to: GPL20182 (Gene Symbol Version)', 'Alternative to: GPL22321 (Alternative CDF [HGU133Plus2_Hs_ENTREZG, Brainarray 20])', 'Alternative to: GPL22945 (alternative)', 'Alternative to: GPL23270 ((Alternative CDF [HGU133Plus2_Hs_ENTREZG, Brainarray 21])', 'Alternative to: GPL23432 (alternative CDF [HGU133Plus2_Hs_ENSG, Brainarray version 18.0.0])', 'Alternative to: GPL25740 (alternative)', 'Alternative to: GPL28718 (Alternative CDF [HGU133Plus2_Hs_ENTREZG, Brainarray 23])', 'Alternative to: GPL29499 (Alternative CDF)'])\n",
      "('data_row_count', ['54675'])\n"
     ]
    }
   ],
   "source": [
    "# What we have here is a complex bespoke object with some finnicky properties, let's unpack that\n",
    "slowpoke = True\n",
    "if slowpoke:\n",
    "    print(type(rawDataGSE10245))\n",
    "    print(rawDataGSE10245.metadata['summary'][0])\n",
    "    print(\"\\nSample GSM258551\")\n",
    "    for metaIndex in rawDataGSE10245.gsms.get(\"GSM258551\").metadata.items():\n",
    "        print(str(metaIndex))\n",
    "    print(\"\\nPlatform GPL570\")\n",
    "    for metaIndex in rawDataGSE10245.gpls.get('GPL570').metadata.items():\n",
    "        print(str(metaIndex))\n",
    "        \n",
    "# Important details include\n",
    "# 1 Platform, meaning that this was all done on the same machine and is therefore comparable\n",
    "# Homo Sapiens as species, informing what references to use for our own analysis\n",
    "# Year of original analysis, informing what references to use for comparative analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "cd22b717-4c01-4355-b30d-84ff074c5fb4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID_REF</th>\n",
       "      <th>VALUE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1007_s_at</td>\n",
       "      <td>9.129905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1053_at</td>\n",
       "      <td>8.034022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>117_at</td>\n",
       "      <td>3.564520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>121_at</td>\n",
       "      <td>4.746490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1255_g_at</td>\n",
       "      <td>2.320698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54670</th>\n",
       "      <td>AFFX-r2-Ec-bioC-5_at</td>\n",
       "      <td>10.730983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54671</th>\n",
       "      <td>AFFX-r2-Ec-bioD-3_at</td>\n",
       "      <td>13.599488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54672</th>\n",
       "      <td>AFFX-r2-Ec-bioD-5_at</td>\n",
       "      <td>13.031726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54673</th>\n",
       "      <td>AFFX-r2-P1-cre-3_at</td>\n",
       "      <td>15.028729</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54674</th>\n",
       "      <td>AFFX-r2-P1-cre-5_at</td>\n",
       "      <td>14.586347</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54675 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     ID_REF      VALUE\n",
       "0                 1007_s_at   9.129905\n",
       "1                   1053_at   8.034022\n",
       "2                    117_at   3.564520\n",
       "3                    121_at   4.746490\n",
       "4                 1255_g_at   2.320698\n",
       "...                     ...        ...\n",
       "54670  AFFX-r2-Ec-bioC-5_at  10.730983\n",
       "54671  AFFX-r2-Ec-bioD-3_at  13.599488\n",
       "54672  AFFX-r2-Ec-bioD-5_at  13.031726\n",
       "54673   AFFX-r2-P1-cre-3_at  15.028729\n",
       "54674   AFFX-r2-P1-cre-5_at  14.586347\n",
       "\n",
       "[54675 rows x 2 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The actual data we want to analyze however is stored in each sample seperately\n",
    "rawDataGSE10245.gsms.get('GSM258551').table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "6b5369bc-f907-4ac1-a2b2-ae61b52cb5ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# So to get a useable dataframe we will join each sample together iteratively. \n",
    "# Another, possibly faster, method would be to check if all of the rows are the same in each file, but this\n",
    "# solution is more generalizable in the case where some rows might be filtered out of some samples\n",
    "dfList = list()\n",
    "# First we will iterate over all of the samples avialable in the object\n",
    "for sampleIndex in rawDataGSE10245.gsms:\n",
    "    # Get the current sample table from the larger object\n",
    "    sampleData = rawDataGSE10245.gsms.get(sampleIndex).table\n",
    "    # get the rownames for assignment to the dataframe\n",
    "    rowNames = sampleData['ID_REF']\n",
    "    # get the actual data for assignment to the dataframe\n",
    "    colData = sampleData['VALUE']\n",
    "    # initialize a temporary dataframe just for the sample and include that samples data with\n",
    "    # the sample number as the column head\n",
    "    sampleDataFrame = pd.DataFrame({sampleIndex:colData})\n",
    "    # reattatch the row names as the index for this temporary dataframe to perserve this sample's\n",
    "    # reference structure on the off chance they are different between samples\n",
    "    sampleDataFrame.index = rowNames\n",
    "    # add this temporary dataframe to the list of all samples' temporary dataframes\n",
    "    dfList.append(sampleDataFrame)\n",
    "# Join all temporary dataframes together by index in one step starting with the first and then \n",
    "# joining all others in the list.\n",
    "rawDataFrameGSE10245 = dfList[0].join(dfList[1:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "73e814b9-1628-4eb5-a7d8-cc66bd92cb6a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                      GSM258551  GSM258552  GSM258553  GSM258554  GSM258555  \\\n",
      "ID_REF                                                                        \n",
      "1007_s_at              9.129905   9.843349   9.730661   9.032165  10.281793   \n",
      "1053_at                8.034022   7.973332   8.834045   7.723965   9.040800   \n",
      "117_at                 3.564520   4.994852   5.066018   4.958580   4.951835   \n",
      "121_at                 4.746490   5.197306   5.234618   6.078180   5.205632   \n",
      "1255_g_at              2.320698   2.248520   2.259504   2.262787   2.207531   \n",
      "...                         ...        ...        ...        ...        ...   \n",
      "AFFX-r2-Ec-bioC-5_at  10.730983   9.719266   9.101115   9.763076   9.651785   \n",
      "AFFX-r2-Ec-bioD-3_at  13.599488  12.847711  12.384142  12.969199  12.924465   \n",
      "AFFX-r2-Ec-bioD-5_at  13.031726  12.250033  11.798363  12.307684  12.243207   \n",
      "AFFX-r2-P1-cre-3_at   15.028729  14.440756  14.439887  14.557363  14.612223   \n",
      "AFFX-r2-P1-cre-5_at   14.586347  14.072366  14.011392  14.162145  14.122751   \n",
      "\n",
      "                      GSM258556  GSM258557  GSM258558  GSM258559  GSM258560  \\\n",
      "ID_REF                                                                        \n",
      "1007_s_at              9.154551   8.063788   8.881805   9.039005  10.643545   \n",
      "1053_at                8.792375   8.093354   7.575620   8.438594   9.142260   \n",
      "117_at                 6.428274   7.364676   4.659326   6.643723   5.368356   \n",
      "121_at                 5.009848   5.533000   5.362637   5.210988   5.195820   \n",
      "1255_g_at              2.322618   2.656929   2.255955   2.266399   2.246563   \n",
      "...                         ...        ...        ...        ...        ...   \n",
      "AFFX-r2-Ec-bioC-5_at   9.716659  10.224866   9.968033   9.698518   9.457238   \n",
      "AFFX-r2-Ec-bioD-3_at  12.996885  13.180944  13.074864  12.782987  12.767953   \n",
      "AFFX-r2-Ec-bioD-5_at  12.369919  12.688711  12.541047  12.103870  12.280146   \n",
      "AFFX-r2-P1-cre-3_at   14.593751  14.697993  14.626500  14.546733  14.523177   \n",
      "AFFX-r2-P1-cre-5_at   14.202961  14.339175  14.292254  14.058803  14.159513   \n",
      "\n",
      "                      ...  GSM258599  GSM258600  GSM258601  GSM258602  \\\n",
      "ID_REF                ...                                               \n",
      "1007_s_at             ...   9.628212  10.130757   9.276419   9.296957   \n",
      "1053_at               ...   9.215567   8.273698   8.790710   8.861625   \n",
      "117_at                ...   5.183919   4.801579   6.161777   5.443126   \n",
      "121_at                ...   5.202508   4.807654   5.326347   5.205004   \n",
      "1255_g_at             ...   2.248392   2.232021   2.246738   2.245739   \n",
      "...                   ...        ...        ...        ...        ...   \n",
      "AFFX-r2-Ec-bioC-5_at  ...   9.163981  10.212906   9.111832   9.888466   \n",
      "AFFX-r2-Ec-bioD-3_at  ...  12.621897  13.323579  12.505348  13.021481   \n",
      "AFFX-r2-Ec-bioD-5_at  ...  12.027085  12.834769  11.794169  12.425024   \n",
      "AFFX-r2-P1-cre-3_at   ...  14.432716  14.779129  14.291955  14.539678   \n",
      "AFFX-r2-P1-cre-5_at   ...  14.062622  14.471579  13.927947  14.149317   \n",
      "\n",
      "                      GSM258603  GSM258604  GSM258605  GSM258606  GSM258607  \\\n",
      "ID_REF                                                                        \n",
      "1007_s_at             10.346701   9.709068   7.672985   9.199080   7.635765   \n",
      "1053_at                8.512193   7.908581   9.299454   8.220959   7.432135   \n",
      "117_at                 4.609810   4.339004   6.384404   6.006534   5.484470   \n",
      "121_at                 5.201123   6.028717   5.220915   5.186452   5.220778   \n",
      "1255_g_at              2.265454   2.301397   2.322604   2.236494   2.243698   \n",
      "...                         ...        ...        ...        ...        ...   \n",
      "AFFX-r2-Ec-bioC-5_at   9.901881   9.814137   9.586620   9.593082  10.066312   \n",
      "AFFX-r2-Ec-bioD-3_at  13.137414  13.131295  12.736871  12.864534  13.248409   \n",
      "AFFX-r2-Ec-bioD-5_at  12.580205  12.448659  12.081561  12.227276  12.616884   \n",
      "AFFX-r2-P1-cre-3_at   14.888295  14.719894  14.534358  14.413188  14.705226   \n",
      "AFFX-r2-P1-cre-5_at   14.425394  14.362869  14.046082  14.083899  14.369607   \n",
      "\n",
      "                      GSM258608  \n",
      "ID_REF                           \n",
      "1007_s_at              8.760082  \n",
      "1053_at                7.596431  \n",
      "117_at                 5.372583  \n",
      "121_at                 5.097535  \n",
      "1255_g_at              2.420731  \n",
      "...                         ...  \n",
      "AFFX-r2-Ec-bioC-5_at  10.343148  \n",
      "AFFX-r2-Ec-bioD-3_at  13.357006  \n",
      "AFFX-r2-Ec-bioD-5_at  12.776164  \n",
      "AFFX-r2-P1-cre-3_at   14.719894  \n",
      "AFFX-r2-P1-cre-5_at   14.371899  \n",
      "\n",
      "[54675 rows x 58 columns]\n"
     ]
    }
   ],
   "source": [
    "# This completes step 1 of Group Project 3 'Load the Data'\n",
    "print(rawDataFrameGSE10245)\n",
    "rawDataFrameGSE10245.to_csv(r'BiorawDataFrameGSE10245.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "4ce2c33c-76eb-4cf9-acc1-3e00471b7eac",
   "metadata": {},
   "outputs": [],
   "source": [
    "df= rawDataFrameGSE10245.T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "160cad8c-bdf3-4f7d-8ed2-5dd06cf99b22",
   "metadata": {},
   "source": [
    "# 2 - Use the sklearn.cluster module to do the clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "60c8d269-11cf-4aae-ba85-d20e56e66573",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Then we need to generate a K-means object using the appropriate number (k) of means\n",
    "# Thanks to Hui for identifying the need for the \"random_state\" parameter to reproduce results.\n",
    "# Thanks to Maya for linking the two for the purposes of testing our script.\n",
    "# When actually using this method you would remove this parameter so as not to risk selecting results.\n",
    "kmeansInit = KMeans(n_clusters = 2,random_state=random_seed)\n",
    "# We give the kmeans object the data in question to generate the cluster points\n",
    "kmeansInit.fit(df)\n",
    "# Finally we ask it to predict the data used to train it and generate the cluster codes for each sample\n",
    "GSE10245Means = kmeansInit.predict(df)\n",
    "# Let's attach these values to a Series labeled with the sample names for the benefit of those who follow\n",
    "GSE10245ClusterCodes = pd.Series(GSE10245Means, index = df.index)\n",
    "GSE10245ClusterCodes.to_csv(r'clusterCodesGSE10245.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "76680509-1f51-4d30-8575-903faa4f9364",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0,\n",
       "       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1,\n",
       "       0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0], dtype=int32)"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GSE10245Means"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "bbfdaaa0-cb5d-40be-a6ab-b9e161cee674",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>ID_REF</th>\n",
       "      <th>1007_s_at</th>\n",
       "      <th>1053_at</th>\n",
       "      <th>117_at</th>\n",
       "      <th>121_at</th>\n",
       "      <th>1255_g_at</th>\n",
       "      <th>1294_at</th>\n",
       "      <th>1316_at</th>\n",
       "      <th>1320_at</th>\n",
       "      <th>1405_i_at</th>\n",
       "      <th>1431_at</th>\n",
       "      <th>...</th>\n",
       "      <th>AFFX-r2-Bs-thr-M_s_at</th>\n",
       "      <th>AFFX-r2-Ec-bioB-3_at</th>\n",
       "      <th>AFFX-r2-Ec-bioB-5_at</th>\n",
       "      <th>AFFX-r2-Ec-bioB-M_at</th>\n",
       "      <th>AFFX-r2-Ec-bioC-3_at</th>\n",
       "      <th>AFFX-r2-Ec-bioC-5_at</th>\n",
       "      <th>AFFX-r2-Ec-bioD-3_at</th>\n",
       "      <th>AFFX-r2-Ec-bioD-5_at</th>\n",
       "      <th>AFFX-r2-P1-cre-3_at</th>\n",
       "      <th>AFFX-r2-P1-cre-5_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM258551</th>\n",
       "      <td>9.129905</td>\n",
       "      <td>8.034022</td>\n",
       "      <td>3.564520</td>\n",
       "      <td>4.746490</td>\n",
       "      <td>2.320698</td>\n",
       "      <td>5.519153</td>\n",
       "      <td>3.339182</td>\n",
       "      <td>2.775395</td>\n",
       "      <td>8.303437</td>\n",
       "      <td>2.981327</td>\n",
       "      <td>...</td>\n",
       "      <td>10.297589</td>\n",
       "      <td>8.857847</td>\n",
       "      <td>9.369037</td>\n",
       "      <td>9.435077</td>\n",
       "      <td>10.936404</td>\n",
       "      <td>10.730983</td>\n",
       "      <td>13.599488</td>\n",
       "      <td>13.031726</td>\n",
       "      <td>15.028729</td>\n",
       "      <td>14.586347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM258552</th>\n",
       "      <td>9.843349</td>\n",
       "      <td>7.973332</td>\n",
       "      <td>4.994852</td>\n",
       "      <td>5.197306</td>\n",
       "      <td>2.248520</td>\n",
       "      <td>5.081258</td>\n",
       "      <td>2.934516</td>\n",
       "      <td>2.617097</td>\n",
       "      <td>9.145519</td>\n",
       "      <td>4.619668</td>\n",
       "      <td>...</td>\n",
       "      <td>8.217455</td>\n",
       "      <td>7.865127</td>\n",
       "      <td>8.116953</td>\n",
       "      <td>8.306683</td>\n",
       "      <td>9.986505</td>\n",
       "      <td>9.719266</td>\n",
       "      <td>12.847711</td>\n",
       "      <td>12.250033</td>\n",
       "      <td>14.440756</td>\n",
       "      <td>14.072366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM258553</th>\n",
       "      <td>9.730661</td>\n",
       "      <td>8.834045</td>\n",
       "      <td>5.066018</td>\n",
       "      <td>5.234618</td>\n",
       "      <td>2.259504</td>\n",
       "      <td>4.657257</td>\n",
       "      <td>3.007192</td>\n",
       "      <td>2.634559</td>\n",
       "      <td>9.159054</td>\n",
       "      <td>2.746464</td>\n",
       "      <td>...</td>\n",
       "      <td>8.713199</td>\n",
       "      <td>6.996104</td>\n",
       "      <td>7.803869</td>\n",
       "      <td>8.029131</td>\n",
       "      <td>9.468096</td>\n",
       "      <td>9.101115</td>\n",
       "      <td>12.384142</td>\n",
       "      <td>11.798363</td>\n",
       "      <td>14.439887</td>\n",
       "      <td>14.011392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM258554</th>\n",
       "      <td>9.032165</td>\n",
       "      <td>7.723965</td>\n",
       "      <td>4.958580</td>\n",
       "      <td>6.078180</td>\n",
       "      <td>2.262787</td>\n",
       "      <td>4.535683</td>\n",
       "      <td>3.167112</td>\n",
       "      <td>3.127495</td>\n",
       "      <td>7.770354</td>\n",
       "      <td>6.212399</td>\n",
       "      <td>...</td>\n",
       "      <td>8.805233</td>\n",
       "      <td>8.221659</td>\n",
       "      <td>8.434090</td>\n",
       "      <td>8.609512</td>\n",
       "      <td>10.064087</td>\n",
       "      <td>9.763076</td>\n",
       "      <td>12.969199</td>\n",
       "      <td>12.307684</td>\n",
       "      <td>14.557363</td>\n",
       "      <td>14.162145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM258555</th>\n",
       "      <td>10.281793</td>\n",
       "      <td>9.040800</td>\n",
       "      <td>4.951835</td>\n",
       "      <td>5.205632</td>\n",
       "      <td>2.207531</td>\n",
       "      <td>3.731919</td>\n",
       "      <td>2.711790</td>\n",
       "      <td>2.707079</td>\n",
       "      <td>6.380928</td>\n",
       "      <td>4.528499</td>\n",
       "      <td>...</td>\n",
       "      <td>8.285092</td>\n",
       "      <td>8.042156</td>\n",
       "      <td>8.174685</td>\n",
       "      <td>8.620098</td>\n",
       "      <td>10.005558</td>\n",
       "      <td>9.651785</td>\n",
       "      <td>12.924465</td>\n",
       "      <td>12.243207</td>\n",
       "      <td>14.612223</td>\n",
       "      <td>14.122751</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 54675 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "ID_REF     1007_s_at   1053_at    117_at    121_at  1255_g_at   1294_at  \\\n",
       "GSM258551   9.129905  8.034022  3.564520  4.746490   2.320698  5.519153   \n",
       "GSM258552   9.843349  7.973332  4.994852  5.197306   2.248520  5.081258   \n",
       "GSM258553   9.730661  8.834045  5.066018  5.234618   2.259504  4.657257   \n",
       "GSM258554   9.032165  7.723965  4.958580  6.078180   2.262787  4.535683   \n",
       "GSM258555  10.281793  9.040800  4.951835  5.205632   2.207531  3.731919   \n",
       "\n",
       "ID_REF      1316_at   1320_at  1405_i_at   1431_at  ...  \\\n",
       "GSM258551  3.339182  2.775395   8.303437  2.981327  ...   \n",
       "GSM258552  2.934516  2.617097   9.145519  4.619668  ...   \n",
       "GSM258553  3.007192  2.634559   9.159054  2.746464  ...   \n",
       "GSM258554  3.167112  3.127495   7.770354  6.212399  ...   \n",
       "GSM258555  2.711790  2.707079   6.380928  4.528499  ...   \n",
       "\n",
       "ID_REF     AFFX-r2-Bs-thr-M_s_at  AFFX-r2-Ec-bioB-3_at  AFFX-r2-Ec-bioB-5_at  \\\n",
       "GSM258551              10.297589              8.857847              9.369037   \n",
       "GSM258552               8.217455              7.865127              8.116953   \n",
       "GSM258553               8.713199              6.996104              7.803869   \n",
       "GSM258554               8.805233              8.221659              8.434090   \n",
       "GSM258555               8.285092              8.042156              8.174685   \n",
       "\n",
       "ID_REF     AFFX-r2-Ec-bioB-M_at  AFFX-r2-Ec-bioC-3_at  AFFX-r2-Ec-bioC-5_at  \\\n",
       "GSM258551              9.435077             10.936404             10.730983   \n",
       "GSM258552              8.306683              9.986505              9.719266   \n",
       "GSM258553              8.029131              9.468096              9.101115   \n",
       "GSM258554              8.609512             10.064087              9.763076   \n",
       "GSM258555              8.620098             10.005558              9.651785   \n",
       "\n",
       "ID_REF     AFFX-r2-Ec-bioD-3_at  AFFX-r2-Ec-bioD-5_at  AFFX-r2-P1-cre-3_at  \\\n",
       "GSM258551             13.599488             13.031726            15.028729   \n",
       "GSM258552             12.847711             12.250033            14.440756   \n",
       "GSM258553             12.384142             11.798363            14.439887   \n",
       "GSM258554             12.969199             12.307684            14.557363   \n",
       "GSM258555             12.924465             12.243207            14.612223   \n",
       "\n",
       "ID_REF     AFFX-r2-P1-cre-5_at  \n",
       "GSM258551            14.586347  \n",
       "GSM258552            14.072366  \n",
       "GSM258553            14.011392  \n",
       "GSM258554            14.162145  \n",
       "GSM258555            14.122751  \n",
       "\n",
       "[5 rows x 54675 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ca805fc-d006-46b2-a253-51c672ede024",
   "metadata": {},
   "source": [
    "# 3 - Create a DataFrame that contains the 58 sample labels as one column, the cluster number as another column and the subtype given in the dataset (adenocarcinoma or squamous cell carcinoma) as another column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "d24b755f-bf0c-4b59-af04-6842dfef3d1d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mbose/miniconda3/lib/python3.9/site-packages/GEOparse/GEOparse.py:545: DtypeWarning: Columns (2) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  table_data = parse_table_data(gpl_soft)\n"
     ]
    }
   ],
   "source": [
    "##########################################################################################\n",
    "#          The following code is a Miguel Limcaoco Production. Enjoy!                    #\n",
    "##########################################################################################\n",
    "\n",
    "#reads in csv file from step 1\n",
    "dataframe = pd.read_csv('clusterCodesGSE10245.csv' , names=['sample_label', 'cluster_number'], skiprows=1)\n",
    "\n",
    "#this declaration might be redundant in the combined notebook\n",
    "rawDataGSE10245 = GEOparse.get_GEO(geo=\"GSE10245\")\n",
    "\n",
    "#creates metadata dictionary and populates with subtype info with keys being sample names since this type of operation is\n",
    "#only being done once i did not make it its own fxn.\n",
    "metadata_dict = {}\n",
    "for gsm_name, gsm in rawDataGSE10245.gsms.items():\n",
    "    for key, value in gsm.metadata.items():\n",
    "        if key == 'characteristics_ch1':\n",
    "            split_value = value[0].split(':')\n",
    "            metadata_dict[gsm_name] = split_value[1]\n",
    "        else:\n",
    "            continue\n",
    "#creating a pandas Series from the collected subtype dictionary and then joining that info to the previous dataframe\n",
    "subtype_series =  pd.Series(metadata_dict)\n",
    "subtype_df= subtype_series.to_frame('subtype')\n",
    "annotated_df = dataframe.join(subtype_df, on='sample_label')\n",
    "\n",
    "#exporting new annotated df\n",
    "annotated_df.to_csv('./clusterCodesAnnotatedGSE10245.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7a85388-633a-4d8f-ac5d-05d0b2ebdef2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "b48710a2-7d3b-4765-a775-e62f35de0347",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "## problem 3 is done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "57f26b4c-33bb-4d8c-996a-9a2465bc6932",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cluster_Number</th>\n",
       "      <th>Tumor_Subtype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM258551</th>\n",
       "      <td>0</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM258552</th>\n",
       "      <td>0</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM258553</th>\n",
       "      <td>1</td>\n",
       "      <td>SCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM258554</th>\n",
       "      <td>0</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM258555</th>\n",
       "      <td>1</td>\n",
       "      <td>SCC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Cluster_Number Tumor_Subtype\n",
       "Sample_Name                              \n",
       "GSM258551                 0            AD\n",
       "GSM258552                 0            AD\n",
       "GSM258553                 1           SCC\n",
       "GSM258554                 0            AD\n",
       "GSM258555                 1           SCC"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "925bdd44-d1e5-4b43-aaac-405c362b929b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for problem 4, Compute and display the accuracy of this clustering.\n",
    "## first set up \"SCC\" = 1 and \"AD\" = 0\n",
    "GSE10245Means\n",
    "GSE10245Means = np.where(GSE10245Means == 0, \"AD\", \"SCC\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "a1db92b9-c20f-4316-9a56-3f495e8b40a9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['AD', 'AD', 'SCC', 'AD', 'SCC', 'AD', 'SCC', 'AD', 'AD', 'AD',\n",
       "       'AD', 'AD', 'SCC', 'AD', 'SCC', 'SCC', 'AD', 'AD', 'AD', 'SCC',\n",
       "       'AD', 'AD', 'AD', 'AD', 'AD', 'AD', 'AD', 'SCC', 'AD', 'AD', 'AD',\n",
       "       'AD', 'SCC', 'AD', 'SCC', 'AD', 'AD', 'AD', 'AD', 'SCC', 'AD',\n",
       "       'AD', 'SCC', 'SCC', 'AD', 'SCC', 'AD', 'AD', 'SCC', 'AD', 'AD',\n",
       "       'AD', 'SCC', 'AD', 'AD', 'AD', 'AD', 'AD'], dtype='<U3')"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GSE10245Means"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "748b0661-ee75-48ca-9824-9799e2225a1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "## want to check the overal accuracy by adding the predicted subtype in the dataframe\n",
    "df_2 = GSE10245ClusterCodes.assign(predicted=GSE10245Means)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "11b003cb-d9f1-4384-badb-846cda9cc9e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cluster_Number</th>\n",
       "      <th>Tumor_Subtype</th>\n",
       "      <th>predicted</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM258551</th>\n",
       "      <td>0</td>\n",
       "      <td>AD</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM258552</th>\n",
       "      <td>0</td>\n",
       "      <td>AD</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM258553</th>\n",
       "      <td>1</td>\n",
       "      <td>SCC</td>\n",
       "      <td>SCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM258554</th>\n",
       "      <td>0</td>\n",
       "      <td>AD</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM258555</th>\n",
       "      <td>1</td>\n",
       "      <td>SCC</td>\n",
       "      <td>SCC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Cluster_Number Tumor_Subtype predicted\n",
       "Sample_Name                                        \n",
       "GSM258551                 0            AD        AD\n",
       "GSM258552                 0            AD        AD\n",
       "GSM258553                 1           SCC       SCC\n",
       "GSM258554                 0            AD        AD\n",
       "GSM258555                 1           SCC       SCC"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_2.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68449a9b-455f-40f6-8bb0-7ef1f9c19454",
   "metadata": {},
   "source": [
    "# 4 - Compute and display the accuracy of this clustering. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "df8fccb2-0e75-4617-886c-6b059f3f8fd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "complete_dataset_acc = accuracy_score(tumor_subtypes, GSE10245Means)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "f8533777-3e58-4b0d-a5b9-52de51a12024",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The accuracy score when testing on all of the data is 0.9310344827586207.\n"
     ]
    }
   ],
   "source": [
    "print(f\"The accuracy score when testing on all of the data is {complete_dataset_acc}.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c1c1d92-64b5-4b1c-ab1f-ea5bad7e6d31",
   "metadata": {},
   "source": [
    "# 5 - Fit the model on half of the data (balanced: 20 AD and 9 SCC) and assign a label to the clusters (AD or SCC), then use the model to predict the cluster/label for the other half of the data."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a3293b7-17a2-4625-bda8-3d05c7b8239a",
   "metadata": {},
   "source": [
    "Lets partition our df into AD and SCC samples. randomly choose 20 AD samples and 9 SCC sample. We will designate these the \"test\" set. The remainder of the samples will be the \"validation\" set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "e6f9fa2f-e695-4967-88e8-87bf59b894ab",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 40 AD samples and 18 SCC samples\n"
     ]
    }
   ],
   "source": [
    "ad = df[df[\"tumor_subtype\"] == \"AD\"]\n",
    "scc = df[df[\"tumor_subtype\"] == \"SCC\"]\n",
    "print(f\"There are {ad.shape[0]} AD samples and {scc.shape[0]} SCC samples\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a0dfc4b-5615-43b7-8b90-2a38f81d8135",
   "metadata": {},
   "source": [
    "We want to use half of the AD samples as our \"training\" set, and the other half as our \"testing\" set. Ditto for the SCC samples. We can use the sklearn.model_selection train_test_split function to do this. test_size indicates the ratio of training samples to testing samples; in this case, the ratio is 50:50 (half)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "3b64ea97-ef41-4492-923c-0c8b08809bf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_ad, test_ad = train_test_split(ad, test_size = 0.5, random_state = random_seed)\n",
    "train_scc, test_scc = train_test_split(scc, test_size = 0.5, random_state = random_seed)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb0c4ef4-e757-4762-a9a2-96c6ead72166",
   "metadata": {},
   "source": [
    "Let's combine our \"training\" set for AD samples with our \"training\" set for SCC samples into a single training set. Do the same to create a single testing set. We also want to remove the tumor subtype annotations, so that we can then use the model to predict these annotations, but we will also save the labels in a seperate variable for testing the accuracy of the model later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "6a04cae0-7909-480f-91f6-439c9b775b67",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Combine AD training samples with SCC training samples to create a single training DF\n",
    "train = pd.concat([train_ad, train_scc])\n",
    "# Save the tumor subtype labels for accuracy testing later\n",
    "train_labels = train[\"tumor_subtype\"].values\n",
    "# drop the tumor subtype labels to use the DF in clustering\n",
    "train = train.drop(\"tumor_subtype\", axis = 1)\n",
    "\n",
    "# Combine AD testing samples with SCC testing samples to create a single testing DF\n",
    "test = pd.concat([test_ad, test_scc])\n",
    "# Save the tumor subtype labels for accuracy testing later\n",
    "test_labels = test[\"tumor_subtype\"].values\n",
    "# drop the tumor subtype labels to use the DF in clustering\n",
    "test = test.drop(\"tumor_subtype\", axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "d0da0cf7-f63a-4f40-9c2b-67ccbdc18138",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>ID_REF</th>\n",
       "      <th>1007_s_at</th>\n",
       "      <th>1053_at</th>\n",
       "      <th>117_at</th>\n",
       "      <th>121_at</th>\n",
       "      <th>1255_g_at</th>\n",
       "      <th>1294_at</th>\n",
       "      <th>1316_at</th>\n",
       "      <th>1320_at</th>\n",
       "      <th>1405_i_at</th>\n",
       "      <th>1431_at</th>\n",
       "      <th>...</th>\n",
       "      <th>AFFX-r2-Bs-thr-M_s_at</th>\n",
       "      <th>AFFX-r2-Ec-bioB-3_at</th>\n",
       "      <th>AFFX-r2-Ec-bioB-5_at</th>\n",
       "      <th>AFFX-r2-Ec-bioB-M_at</th>\n",
       "      <th>AFFX-r2-Ec-bioC-3_at</th>\n",
       "      <th>AFFX-r2-Ec-bioC-5_at</th>\n",
       "      <th>AFFX-r2-Ec-bioD-3_at</th>\n",
       "      <th>AFFX-r2-Ec-bioD-5_at</th>\n",
       "      <th>AFFX-r2-P1-cre-3_at</th>\n",
       "      <th>AFFX-r2-P1-cre-5_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM258598</th>\n",
       "      <td>9.871776</td>\n",
       "      <td>7.469301</td>\n",
       "      <td>3.583423</td>\n",
       "      <td>5.736595</td>\n",
       "      <td>2.225399</td>\n",
       "      <td>4.302100</td>\n",
       "      <td>2.963642</td>\n",
       "      <td>2.599405</td>\n",
       "      <td>6.248514</td>\n",
       "      <td>4.227329</td>\n",
       "      <td>...</td>\n",
       "      <td>8.912983</td>\n",
       "      <td>8.379795</td>\n",
       "      <td>8.786520</td>\n",
       "      <td>8.760903</td>\n",
       "      <td>10.232256</td>\n",
       "      <td>9.824295</td>\n",
       "      <td>13.162891</td>\n",
       "      <td>12.446226</td>\n",
       "      <td>14.703199</td>\n",
       "      <td>14.231290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM258554</th>\n",
       "      <td>9.032165</td>\n",
       "      <td>7.723965</td>\n",
       "      <td>4.958580</td>\n",
       "      <td>6.078180</td>\n",
       "      <td>2.262787</td>\n",
       "      <td>4.535683</td>\n",
       "      <td>3.167112</td>\n",
       "      <td>3.127495</td>\n",
       "      <td>7.770354</td>\n",
       "      <td>6.212399</td>\n",
       "      <td>...</td>\n",
       "      <td>8.805233</td>\n",
       "      <td>8.221659</td>\n",
       "      <td>8.434090</td>\n",
       "      <td>8.609512</td>\n",
       "      <td>10.064087</td>\n",
       "      <td>9.763076</td>\n",
       "      <td>12.969199</td>\n",
       "      <td>12.307684</td>\n",
       "      <td>14.557363</td>\n",
       "      <td>14.162145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM258586</th>\n",
       "      <td>8.757455</td>\n",
       "      <td>6.958085</td>\n",
       "      <td>5.339252</td>\n",
       "      <td>5.193266</td>\n",
       "      <td>5.535001</td>\n",
       "      <td>5.132130</td>\n",
       "      <td>2.973675</td>\n",
       "      <td>2.607981</td>\n",
       "      <td>8.864519</td>\n",
       "      <td>2.723080</td>\n",
       "      <td>...</td>\n",
       "      <td>10.209200</td>\n",
       "      <td>8.472484</td>\n",
       "      <td>8.856784</td>\n",
       "      <td>8.997555</td>\n",
       "      <td>10.459636</td>\n",
       "      <td>10.111940</td>\n",
       "      <td>13.365125</td>\n",
       "      <td>12.720290</td>\n",
       "      <td>14.852756</td>\n",
       "      <td>14.426513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM258577</th>\n",
       "      <td>8.711734</td>\n",
       "      <td>8.604424</td>\n",
       "      <td>6.845243</td>\n",
       "      <td>5.207483</td>\n",
       "      <td>2.694211</td>\n",
       "      <td>4.768066</td>\n",
       "      <td>3.032224</td>\n",
       "      <td>2.657855</td>\n",
       "      <td>8.724363</td>\n",
       "      <td>3.388218</td>\n",
       "      <td>...</td>\n",
       "      <td>9.527837</td>\n",
       "      <td>6.362626</td>\n",
       "      <td>6.053217</td>\n",
       "      <td>6.575966</td>\n",
       "      <td>8.208702</td>\n",
       "      <td>8.028383</td>\n",
       "      <td>11.462046</td>\n",
       "      <td>10.570192</td>\n",
       "      <td>13.575853</td>\n",
       "      <td>13.007838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM258574</th>\n",
       "      <td>8.914079</td>\n",
       "      <td>8.232630</td>\n",
       "      <td>6.634602</td>\n",
       "      <td>5.086281</td>\n",
       "      <td>2.212639</td>\n",
       "      <td>5.626913</td>\n",
       "      <td>2.949485</td>\n",
       "      <td>2.578616</td>\n",
       "      <td>9.465973</td>\n",
       "      <td>4.645915</td>\n",
       "      <td>...</td>\n",
       "      <td>9.503383</td>\n",
       "      <td>7.862734</td>\n",
       "      <td>8.362031</td>\n",
       "      <td>8.540746</td>\n",
       "      <td>10.206568</td>\n",
       "      <td>9.676948</td>\n",
       "      <td>13.081193</td>\n",
       "      <td>12.531710</td>\n",
       "      <td>14.738462</td>\n",
       "      <td>14.298989</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 54675 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "ID_REF     1007_s_at   1053_at    117_at    121_at  1255_g_at   1294_at  \\\n",
       "GSM258598   9.871776  7.469301  3.583423  5.736595   2.225399  4.302100   \n",
       "GSM258554   9.032165  7.723965  4.958580  6.078180   2.262787  4.535683   \n",
       "GSM258586   8.757455  6.958085  5.339252  5.193266   5.535001  5.132130   \n",
       "GSM258577   8.711734  8.604424  6.845243  5.207483   2.694211  4.768066   \n",
       "GSM258574   8.914079  8.232630  6.634602  5.086281   2.212639  5.626913   \n",
       "\n",
       "ID_REF      1316_at   1320_at  1405_i_at   1431_at  ...  \\\n",
       "GSM258598  2.963642  2.599405   6.248514  4.227329  ...   \n",
       "GSM258554  3.167112  3.127495   7.770354  6.212399  ...   \n",
       "GSM258586  2.973675  2.607981   8.864519  2.723080  ...   \n",
       "GSM258577  3.032224  2.657855   8.724363  3.388218  ...   \n",
       "GSM258574  2.949485  2.578616   9.465973  4.645915  ...   \n",
       "\n",
       "ID_REF     AFFX-r2-Bs-thr-M_s_at  AFFX-r2-Ec-bioB-3_at  AFFX-r2-Ec-bioB-5_at  \\\n",
       "GSM258598               8.912983              8.379795              8.786520   \n",
       "GSM258554               8.805233              8.221659              8.434090   \n",
       "GSM258586              10.209200              8.472484              8.856784   \n",
       "GSM258577               9.527837              6.362626              6.053217   \n",
       "GSM258574               9.503383              7.862734              8.362031   \n",
       "\n",
       "ID_REF     AFFX-r2-Ec-bioB-M_at  AFFX-r2-Ec-bioC-3_at  AFFX-r2-Ec-bioC-5_at  \\\n",
       "GSM258598              8.760903             10.232256              9.824295   \n",
       "GSM258554              8.609512             10.064087              9.763076   \n",
       "GSM258586              8.997555             10.459636             10.111940   \n",
       "GSM258577              6.575966              8.208702              8.028383   \n",
       "GSM258574              8.540746             10.206568              9.676948   \n",
       "\n",
       "ID_REF     AFFX-r2-Ec-bioD-3_at  AFFX-r2-Ec-bioD-5_at  AFFX-r2-P1-cre-3_at  \\\n",
       "GSM258598             13.162891             12.446226            14.703199   \n",
       "GSM258554             12.969199             12.307684            14.557363   \n",
       "GSM258586             13.365125             12.720290            14.852756   \n",
       "GSM258577             11.462046             10.570192            13.575853   \n",
       "GSM258574             13.081193             12.531710            14.738462   \n",
       "\n",
       "ID_REF     AFFX-r2-P1-cre-5_at  \n",
       "GSM258598            14.231290  \n",
       "GSM258554            14.162145  \n",
       "GSM258586            14.426513  \n",
       "GSM258577            13.007838  \n",
       "GSM258574            14.298989  \n",
       "\n",
       "[5 rows x 54675 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check out what one of these DFs looks like to make sure everything's good\n",
    "train.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c32cd14-6f2e-46ac-9bdc-f1375d485fdb",
   "metadata": {},
   "source": [
    "Now, we can train the model using the training set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "a1ac84b7-161e-473d-97ba-0f2d7afc9384",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Train the model\n",
    "kmeansInit.fit(train)\n",
    "\n",
    "# Use the trained model to predict tumor states in the training set\n",
    "train_kmeans = kmeansInit.predict(train)\n",
    "# Use the trained model to predict tumor states in the testing set\n",
    "test_kmeans = kmeansInit.predict(test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "94f988b8-ff50-4df0-9788-6a7592401317",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1,\n",
       "       1, 1, 1, 1, 1, 1, 0], dtype=int32)"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check out what the clusters look like\n",
    "train_kmeans "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52fafbc0-6a5e-4290-84ff-92dafd21e672",
   "metadata": {},
   "source": [
    "We will map the 0 cluster to AD and the 1 cluster to SSC."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "fd719a88-97df-4479-9995-229a4a9f3b0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_kmeans = np.where(train_kmeans == 0, \"AD\", \"SCC\")\n",
    "test_kmeans = np.where(test_kmeans == 0, \"AD\", \"SCC\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "ae29e9a5-ed32-4a56-acec-ae95050d957d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['AD', 'AD', 'AD', 'AD', 'AD', 'AD', 'AD', 'AD', 'SCC', 'AD', 'AD',\n",
       "       'AD', 'AD', 'AD', 'AD', 'AD', 'AD', 'AD', 'AD', 'AD', 'SCC', 'SCC',\n",
       "       'SCC', 'SCC', 'SCC', 'SCC', 'SCC', 'SCC', 'AD'], dtype='<U3')"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_kmeans"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e140252f-da43-4dfe-8e0b-962e4fc1a108",
   "metadata": {},
   "source": [
    "Neat. Step 5 completed! Now for step 6 (computing and displaying the accuracy of the model for the training and testing data). We can do this by using sklearn.metrics functions accuracy_score to compare the known labels (stored in variables earlier) to the cluster predictions.\n",
    "Accuracy score for the training data set:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94b07ae0-b2e2-4e6c-880e-cc4416eb4bed",
   "metadata": {},
   "source": [
    "# 6 - Compute and display the accuracy of the model for the prediction results (testing data)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "fb6d87a3-6166-417d-9b92-e79b3f6e9a42",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The accuracy score when training on half of the data is 0.9310344827586207.\n"
     ]
    }
   ],
   "source": [
    "print(f\"The accuracy score when training on half of the data is {accuracy_score(train_labels, train_kmeans)}.\")\n",
    "training_acc = accuracy_score(train_labels, train_kmeans)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c016eb8c-1191-4c46-a55f-0bff0d81134a",
   "metadata": {},
   "source": [
    "Accuracy score for the testing data set:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "d927544f-7315-44cf-9f35-4d6915dbbddc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The accuracy score when testing on half of the data is 0.896551724137931.\n"
     ]
    }
   ],
   "source": [
    "print(f\"The accuracy score when testing on half of the data is {accuracy_score(test_labels, test_kmeans)}.\")\n",
    "testing_acc = accuracy_score(test_labels, train_kmeans)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54adc3d0-c66b-4c18-8468-72cf92d9adf0",
   "metadata": {},
   "source": [
    "# 7 - Plot the 3 accuracy values: for the model when training on all the data, for the model when trining on half of the data, for the model when testing on the half of the data (test data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9a7a8d2-1680-4806-a282-a2470787b649",
   "metadata": {},
   "source": [
    "The three numbers we need to plot are 0.9310344827586207 for  complete dataset, 0.9310344827586207 for training dataset, and 0.896551724137931 for testing dataset\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "6a46ae0e-7353-4627-a18e-638537ec9407",
   "metadata": {},
   "outputs": [],
   "source": [
    "## create a bar chart and \n",
    "## assign the name for each bar \n",
    "names = ['complete_dataset','training_half_dataset','testing_half_dataset']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "7e12e0c5-d946-4ad6-a841-334bd9c7eb59",
   "metadata": {},
   "outputs": [],
   "source": [
    "## assigne the accuracy_score to each bar \n",
    "values = [complete_dataset_acc, training_acc, testing_acc]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "972a2e78-f48f-4e52-b4b2-5064b8a5dbe1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0.5, 1.0, 'Plot the 3 accuracy values')"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYIAAAEJCAYAAACZjSCSAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8rg+JYAAAACXBIWXMAAAsTAAALEwEAmpwYAAAd2UlEQVR4nO3de7xVZb3v8c9X0IRA0MC2IQgpZtoxd6JpF6WbqeVG9840tcJSIzO1nR7d5TFtu0+2dR/bvbyQepSTZqhpaoZ4KyVvBSioYBqhBEmFBl5QM+R3/nieJYPJnGuNtVhjLRfj+3695muNyzOe8RtzjDl/YzzPmGMpIjAzs/raqLcDMDOz3uVEYGZWc04EZmY150RgZlZzTgRmZjXnRGBmVnNOBNYpku6SdFRfrd+6TtIUSWf1dhzW/ZwIbB2SnpL0sqQXJf1Z0uWSBnWyjtGSQlL/dsqcIenK9Y+4Zf1XSloq6XlJTzjBmDXnRGCtHBARg4D3ALsBp/VyPF3xHWB0RGwG/BNwlqRdezmmptpLmGZVcyKwdkXEH4FbgHc1zpO0kaTTJC2S9BdJP5Q0JM+ekf+uyFcWezYsuy/wDeCQPH9uYfY2ku6V9IKk2yQNKyy3h6T7JK2QNFfS+HZinxcRf2sbza9tm5WVtK2kX0h6VtIzkn4kaWhh/khJ10talsucX5h3tKTHcrzzJb0nTw9J2xXKvd60Imm8pCWSTpH0J+BySZtLujmvY3ke3rqw/Bb56uzpPP+GPP1RSQcUym2ct2GXJtv5mKRPFsb757JtMV8r6U+SnpM0Q9JOLd6viZLuaZj2+vZKepOkcyX9IV9VTpY0IM8blrdthaS/SvqVJH8X9SK/+dYuSSOB/YGHmsyemF8fAt4ODALaviD3yn+HRsSgiLi/uGBETAf+N3B1nv/uwuzDgCOBLYFNgJNyLCOAnwNnAVvk6ddJGt5O/BdKegn4LbAUmNaqKOkK4m3AO4GRwBm5jn7AzcAiYDQwApia5x2cy30OaLvyeLZVPA3+IW/HNsAxpM/j5Xl8FPAya95PgCuAgcBOpPfmvDz9h8ARhXL7A0sjYk6Tdf4Y+Exh/OPAMxHxYB6/BRib638Q+FHJbWn0XWB7YBdgO9J7dnqe93VgCTAceCvphMDPuulNEeGXX2u9gKeAF4EVpC+/C4EBed5dwFF5+E7g2MJy7wD+DvQnfWEG0L+d9ZwBXNkw7S7gtML4scD0PHwKcEVD+VuBz3ewPf2AD5CatzYu+R4cCDyUh/cEljXblrz+E1rUEcB2hfEpwFl5eDzwKrBpOzHsAizPw1sBq4HNm5R7G/ACsFke/wnwP1vUuV0uOzCP/wg4vUXZoXkbhjSJfyJwT7PtJSXVlcC2hXl7Ak/m4W8DNxbfG7969+UrAmvlwIgYGhHbRMSxEfFykzJvIyWKNotISeCt67nuPxWGXyJdaUA6Uz44NymskLSC9AW/VXuVRcRrEXEPsDXw5WZlJG0paaqkP0p6HrgSaGuSGgksiohVTRYdCfy+5HY1WhYRrxRiGCjpB7mp7XlS89rQfEUyEvhrRCxvsn1PA/cC/5Kbs/ajxZl8RCwAHgMOkDSQdAVzVV5/P0lnS/p9Xv9TebFhzepqx3DSlcvswn6anqcDnAMsAG6TtFDSqZ2s37qZE4Gtj6dJX85tRgGrgD9T7lK/s80Bi0lXBEMLrzdHxNkll+9Piz4CUrNQADtH6lw+gnRm27beUS06dBe3U+dLpC/ENv/QML9x+79Ouqp6b46hrXlNeT1bFPstGvy/HPPBwP2R+nZaaWsemgDMz8kBUpPcBOCjwBDSVV3b+hutpLBtkorb9gypWWunwn4aEunmAyLihYj4ekS8HTgA+FdJH2knXquYE4Gtjx8DX5M0Run20rY2/1WkppTVpL6DVv4MjO5ER+GVpDPZj+ez101zp+vWjQXzGf6hkgblsh8nffn9okXdg8nNYbkv4uTCvN+Q+hfOlvTmvN7353mXAidJ2lXJdpLakuMc4LC8/n2BvTvYvsGkL9AVkrYAvtU2IyKWktrvL8ydyhtL2quw7A2kO7xOIPUZtGcqsA/p6uiqhvX/jdTHMZC0P1uZC+wkaRdJm5L7U3Ksq4FLgPMkbQmpfyfvAyR9Mr9PAp4HXssv6yVOBLY+LiN1YM4AngReAb4KEBEvAf8B3JubB/Zosvy1+e+zkh5sMn8tEbGYdMb6DVKiWUz6wm52HAfpi24JsBw4FzgxIm5sUf2ZpC/S50gd0tcX1vsa6cx1O+APuc5D8rxr83ZeRWp7v4HUAQzpS/kAUl/L4Xlee74HDCCdUT9Aak4p+iypD+a3wF+AEwsxvgxcB4wpxt5MTir3A+8Dri7M+iGpee+PwPwcQ6s6niC19d8B/A64p6HIKaTmnwdyM9MdpKsdSJ3Rd5AS7/3AhRFxV3sxW7UU4c56sw2BpNOB7SPiiA4LmxX4RyxmG4DclPRF0lWDWae4acisj5N0NKmZ7JaImNFRebNGbhoyM6s5XxGYmdVcn+sjGDZsWIwePbq3wzAz61Nmz579TEQ0fRxLn0sEo0ePZtasWb0dhplZnyJpUat5bhoyM6s5JwIzs5pzIjAzqzknAjOzmnMiMDOrOScCM7OacyIwM6s5JwIzs5pzIjAzq7k+98vi9TH61J/3dggbrKfO/kQl9XqfVcf7rO+pap/5isDMrOacCMzMas6JwMys5pwIzMxqzonAzKzmnAjMzGrOicDMrOacCMzMas6JwMys5pwIzMxqzonAzKzmnAjMzGrOicDMrOacCMzMas6JwMys5pwIzMxqzonAzKzmnAjMzGrOicDMrOacCMzMas6JwMys5pwIzMxqzonAzKzmnAjMzGrOicDMrOYqTQSS9pX0uKQFkk5tMn+IpJ9JmitpnqQjq4zHzMzWVVkikNQPuADYD9gR+IykHRuKfQWYHxHvBsYD/yVpk6piMjOzdVV5RbA7sCAiFkbEq8BUYEJDmQAGSxIwCPgrsKrCmMzMrEGViWAEsLgwviRPKzofeCfwNPAIcEJErG6sSNIxkmZJmrVs2bKq4jUzq6UqE4GaTIuG8Y8Dc4C3AbsA50vabJ2FIi6OiHERMW748OHdHaeZWa1VmQiWACML41uTzvyLjgSuj2QB8CSwQ4UxmZlZgyoTwUxgrKQxuQP4UOCmhjJ/AD4CIOmtwDuAhRXGZGZmDfpXVXFErJJ0HHAr0A+4LCLmSZqU508G/h2YIukRUlPSKRHxTFUxmZnZuipLBAARMQ2Y1jBtcmH4aWCfKmMwM7P2+ZfFZmY150RgZlZzTgRmZjXnRGBmVnNOBGZmNedEYGZWc04EZmY150RgZlZzTgRmZjXnRGBmVnNOBGZmNedEYGZWc04EZmY150RgZlZzTgRmZjXnRGBmVnNOBGZmNedEYGZWc04EZmY150RgZlZzTgRmZjXnRGBmVnNOBGZmNedEYGZWc04EZmY150RgZlZzTgRmZjVXKhFI2kbSR/PwAEmDqw3LzMx6SoeJQNLRwE+AH+RJWwM3VBiTmZn1oDJXBF8B3g88DxARvwO2rDIoMzPrOWUSwd8i4tW2EUn9gaguJDMz60llEsHdkr4BDJD0MeBa4GfVhmVmZj2lTCI4BVgGPAJ8CZgGnFZlUGZm1nP6tzdT0kbAwxHxLuCSngnJzMx6UrtXBBGxGpgraVQPxWNmZj2sTNPQVsA8SXdKuqntVaZySftKelzSAkmntigzXtIcSfMk3d2Z4M3MbP212zSUndmViiX1Ay4APgYsAWZKuiki5hfKDAUuBPaNiD9I8m2pZmY9rMMrgoi4G/gtMDi/HsvTOrI7sCAiFubbT6cCExrKHAZcHxF/yOv6S2eCNzOz9Vfml8WfBn4DHAx8Gvi1pE+VqHsEsLgwviRPK9oe2FzSXZJmS/pcixiOkTRL0qxly5aVWLWZmZVVpmnom8BubWfrkoYDd5AeO9EeNZnW+EO0/sCuwEeAAcD9kh6IiCfWWijiYuBigHHjxvnHbGZm3ahMItioocnmWcp1Mi8BRhbGtwaeblLmmYhYCayUNAN4N/AEZmbWI8p8oU+XdKukiZImAj8Hbimx3ExgrKQxkjYBDgUa7za6EfigpP6SBgLvBR4rH76Zma2vDq8IIuJkSf8MfIDU3HNxRPy0xHKrJB0H3Ar0Ay6LiHmSJuX5kyPiMUnTgYeB1cClEfHoemyPmZl1UoeJQNIYYFpEXJ/HB0gaHRFPdbRsREwjPZKiOG1yw/g5wDmdCdrMzLpPmaaha0ln621ey9PMzGwDUCYR9C8+hjoPb1JdSGZm1pPKJIJlkv6pbUTSBOCZ6kIyM7OeVOb20UnAjySdT+osXgw0/eGXmZn1PWXuGvo9sIekQYAi4oXqwzIzs55S5hETJ0jaDFgJnCfpQUn7VB+amZn1hDJ9BF+IiOeBfUj/tP5I4OxKozIzsx5TJhG0PTNof+DyiJhL8+cImZlZH1QmEcyWdBspEdwqaTBr/67AzMz6sDJ3DX0R2AVYGBEvSXoLqXkIAEk7RcS8iuIzM7OKlblraDXwYGH8WdITSNtcAbyn+0MzM7OeUKZpqCPuLzAz68O6IxH4H8WYmfVh3ZEIzMysD+uORPBqx0XMzOyNqswvi6+T9AlJTctGxB7dH5aZmfWUMlcEFwGHAb+TdLakHSqOyczMelCHiSAi7oiIw0m3iD4F3C7pPklHStq46gDNzKxapfoI8o/IJgJHAQ8B/01KDLdXFpmZmfWIMv+z+HpgB9IPxw6IiKV51tWSZlUZnJmZVa/MIybOj4hfNJsREeO6OR4zM+thZZqG3ilpaNuIpM0lHVtdSGZm1pPKJIKjI2JF20hELAeOriwiMzPrUWUSwUaSXn+ekKR+wCbVhWRmZj2pTB/BrcA1kiaTnis0CZheaVRmZtZjyiSCU4AvAV8mPWn0NuDSKoMyM7OeU/b/EVyUX2ZmtoEp8zuCscB3gB2BTdumR8TbK4zLzMx6SJnO4stJVwOrgA8BPyT9uMzMzDYAZRLBgIi4E1BELIqIM4APVxuWmZn1lDKdxa/kR1D/TtJxwB+BLasNy8zMekqZK4ITgYHA8cCuwBHA5yuMyczMelC7VwT5x2OfjoiTgReBI3skKjMz6zHtXhFExGvArsVfFpuZ2YalTB/BQ8CNkq4FVrZNjIjrK4vKzMx6TJlEsAXwLGvfKRSAE4GZ2QagzC+Lu9wvIGlf0n8z6wdcGhFntyi3G/AAcEhE/KSr6zMzs84r88viy0lXAGuJiC90sFw/4ALgY8ASYKakmyJifpNy3yU93M7MzHpYmaahmwvDmwIHAU+XWG53YEFELASQNBWYAMxvKPdV4DpgtxJ1mplZNyvTNHRdcVzSj4E7StQ9AlhcGF8CvLehrhGkxPJh2kkEko4BjgEYNWpUiVWbmVlZZX5Q1mgsUObbuNktp41NTN8DTsm3qbYUERdHxLiIGDd8+PByUZqZWSll+gheYO0v8D+R/kdBR5YAIwvjW7Nuk9I4YGr+mcIwYH9JqyLihhL1m5lZNyjTNDS4i3XPBMZKGkN6PtGhwGENdY9pG5Y0BbjZScDMrGd12DQk6SBJQwrjQyUd2NFyEbEKOI50N9BjwDURMU/SJEmT1iNmMzPrRmXuGvpWRPy0bSQiVkj6FnBDRwtGxDRgWsO0yS3KTiwRi5mZdbMyncXNypRJIGZm1geUSQSzJP0fSdtKeruk84DZVQdmZmY9o0wi+CrwKnA1cA3wMvCVKoMyM7OeU+auoZXAqT0Qi5mZ9YIydw3dLmloYXxzSX4ukJnZBqJM09CwiFjRNhIRy/H/LDYz22CUSQSrJb3+SAlJo2nyNFIzM+ubytwG+k3gHkl35/G9yA+AMzOzvq9MZ/F0SeNIX/5zgBtJdw6ZmdkGoMxD544CTiA9NG4OsAdwP2v/60ozM+ujyvQRnED6XwGLIuJDwD8CyyqNyszMekyZRPBKRLwCIOlNEfFb4B3VhmVmZj2lTGfxkvw7ghuA2yUtp9y/qjQzsz6gTGfxQXnwDEm/BIYA0yuNyszMekynniIaEXd3XMrMzPqSrvzPYjMz24A4EZiZ1ZwTgZlZzTkRmJnVnBOBmVnNORGYmdWcE4GZWc05EZiZ1ZwTgZlZzTkRmJnVnBOBmVnNORGYmdWcE4GZWc05EZiZ1ZwTgZlZzTkRmJnVnBOBmVnNORGYmdWcE4GZWc05EZiZ1VyliUDSvpIel7RA0qlN5h8u6eH8uk/Su6uMx8zM1lVZIpDUD7gA2A/YEfiMpB0bij0J7B0ROwP/DlxcVTxmZtZclVcEuwMLImJhRLwKTAUmFAtExH0RsTyPPgBsXWE8ZmbWRJWJYASwuDC+JE9r5YvALc1mSDpG0ixJs5YtW9aNIZqZWZWJQE2mRdOC0odIieCUZvMj4uKIGBcR44YPH96NIZqZWf8K614CjCyMbw083VhI0s7ApcB+EfFshfGYmVkTVV4RzATGShojaRPgUOCmYgFJo4Drgc9GxBMVxmJmZi1UdkUQEaskHQfcCvQDLouIeZIm5fmTgdOBtwAXSgJYFRHjqorJzMzWVWXTEBExDZjWMG1yYfgo4KgqYzAzs/b5l8VmZjXnRGBmVnNOBGZmNedEYGZWc04EZmY150RgZlZzTgRmZjXnRGBmVnNOBGZmNedEYGZWc04EZmY150RgZlZzTgRmZjXnRGBmVnNOBGZmNedEYGZWc04EZmY150RgZlZzTgRmZjXnRGBmVnNOBGZmNedEYGZWc04EZmY150RgZlZzTgRmZjXnRGBmVnNOBGZmNedEYGZWc04EZmY150RgZlZzTgRmZjXnRGBmVnNOBGZmNedEYGZWc04EZmY150RgZlZzlSYCSftKelzSAkmnNpkvSd/P8x+W9J4q4zEzs3VVlggk9QMuAPYDdgQ+I2nHhmL7AWPz6xjgoqriMTOz5qq8ItgdWBARCyPiVWAqMKGhzATgh5E8AAyVtFWFMZmZWYP+FdY9AlhcGF8CvLdEmRHA0mIhSceQrhgAXpT0ePeG+oY1DHimt4MoQ9/t7QjeMLzP+pY+s79gvffZNq1mVJkI1GRadKEMEXExcHF3BNWXSJoVEeN6Ow4rz/usb/H+SqpsGloCjCyMbw083YUyZmZWoSoTwUxgrKQxkjYBDgVuaihzE/C5fPfQHsBzEbG0sSIzM6tOZU1DEbFK0nHArUA/4LKImCdpUp4/GZgG7A8sAF4Cjqwqnj6qds1hGwDvs77F+wtQxDpN8mZmViP+ZbGZWc05EZiZ1ZwTgZlZzTkRVETSFEmf6qDMRElv64Z1jZb0aIkyh63vuprUe6KkgV1cdqikY7uw3DRJQzso821JH+1KXC3q6/A9brLMGZJOysM7SJoj6SFJ23Zm2XbKHNjksS3rpexx0tV9l5dd65gpsz87Wf9ESed3cpm7JI3LwwdLekzSLzu7bDtluvw5aafO8ZLe1x11ORH0ronAeieCkkYD3Z4IgBOBrh7gQ4F1vkzyc6paioj9I2JFB2VOj4g7uhhXFQ4EboyIf4yI33djnd2aCCh/nAylyb4r6UQKx0yZ/dnDvggcGxEf6sY6T6Trn5NWxgPdkgiIiFq9gM8BDwNzgStIP7u+M0+7ExiVy00hPQTvl8BCYG/gMuAxYEqhvheB/wIezMsPLyz/qTy8K3A3MJt0O+1WwKfyso8Dc4ABzcq1sx275m24HzgHeDRPHw38KsfzIPC+PP0B4Lm8rq+1U24rYEYu9yjwwTx9n7yuB4FrgUHA8cCrwCPAL7uwL6YCL+d1zczv9VXA/Dz/hvxezAOOKSz3FOnRAKPz/rgkl7kNGNDk/X8KODPH/giwQ54+HLg9T/8BsAgY1iLW9tZ1dI5/LnAdMDBPPwM4iXSL9J+AP7b3PgHfzMfDHcCPgZNa1U/6Avgr8GR+/7ZtJ46D876cC8zI0/qRjpuZpGP/S82Ok5L77hzg5EJdZ+YybwZ+ntf7KHBIs2Om5P7cLde91vHeIraJwPXAdOB3wH8W5l0EzMr1n1mYfhcwDjidNZ/Lc1rUPyBv/8PA1cCvgXGt6m+xza3iOBuYn+s+t3CcXpff35nA+/N71XZMzSF/Trv8vdgbX8a99QJ2yjt4WB7fAvgZ8Pk8/gXghljzRTKV9BiMCcDzwP8gXUXNBnbJ5QI4PA+fDpxfWP5TwMbAfaxJEIeQflPx+sGXh1uWa7EtDwN75+FiIhgIbJqHxwKz8vB44ObC8q3KfR34Zh7uBwwmfUhnAG/O008BTi9+iLu4P0YX4h4PrATGFOZvUfjgPQq8pbjOvPyqwr64Bjii+P4Xyn81Dx8LXJqHzwf+LQ/vm/dle4mg1breUih3VmFdZ7Dmy/z14Rb170r6ohgIbEb6bc1JHdT/+jZ2UO4RYEQeHpr/HgOcloffRPpSGtN4nJTcd/uQ7scX6fNxM7AX8C/AJYVlhjQ7Zkruz0dZc7JyNh0ngoXAEGBTUoIf2XBM9SN9/nZu8ll8fbhF/f/Kms/wzjnmcR3U37jN65QjfR89zprb+tv21VXAB/LwKOCxMsdUZ15VPmvojejDwE8i4hmAiPirpD2Bf87zrwD+s1D+ZxERkh4B/hwRjwBImkc6aOcAq0lnBQBXks5Eit4BvAu4XRKkHd/s19NlyyFpCOkgubsQ9355eGPgfEm7AK8B2zd/K1qWmwlcJmljUlKcI2lvUhPEvTm2TUhnZt3tNxHxZGH8eEkH5eGRpIT1bMMyT0bEnDw8m7Rfmrm+UKZtf38AOAggIqZLWt5BfK3W9S5JZ5GaSwaRruY664PATyPiJQBJxV/hl62/Vbl7gSmSrmHN+7APsHOhH2sI6f19tQux75NfD+XxQbmuXwHnSvouKbn8qkRd67zHuf9gcETcl6dfBXyyg3rujIjnACTNJ135LwY+nR9i2Z909bsj6aSqM/YCvg8QEQ9LKi5ftv5m5eYDrwCXSvo5KaECfBTYMX/2ADaTNLiTMberbolANHmoXYPi/L/lv6sLw23jrd67xvoFzIuIPUvEVqZcW9lW2/E14M/Au0lnZ690plxEzJC0F/AJ4ApJ5wDLgdsj4jMlYlsfK9sGJI0nfQD2jIiXJN1FOrtrVNwvr5GuHpr5W6FM275r9tDD9rRa1xTgwIiYK2ki6ay6K1rt07L1Ny0XEZMkvZe0T+fk5C/SFcNaSSW/750l4DsR8YN1Zki7kprGviPptoj4dgd1NXuPO7ufmtXTX9IYUlPdbhGxXNIUmh9TZayzr8rW36pcpKcx7A58hPRInuNIJ68bkT4HLzfU08XQ11W3zuI7SZn4LQCStiA1xxya5x8O3NPJOjciNQFB6mRrXP5xYHi+8kDSxpJ2yvNeIDW9dFRuLZE61p6T9IFC3G2GAEsjYjXwWdKVReO6WpaTtA3wl4i4BPi/wHtI7cbvl7RdLjNQ0vYt6u2M9pYdAizPSWAHYI8urqM99wCfBpC0D7B5F+sZDCzNV1GHd1S4hRnAQZIG5LO9A0rU3/j+NS0naduI+HVEnE565PJI0tXCl3NZJG0v6c1N6mylWO5W4AuSBuW6RkjaMt8R91JEXAmcSzqWmsXdrohYDryg9DwyWPN57azNSCcbz0l6K2uuojtrBvn9lfQuUrNOR/UXt7lpufz+DYmIaaTO5V1y+dtISYFcrm36+nz21lKrK4JIzzr6D+BuSa+RLmWPJzWFnAwso/PPO1oJ7CRpNqmT7ZCGdb6aL7+/n5t0+gPfI3USTQEmS3oZ2JOUUJqVa+bIHPdLrN1UcCFwnaSDSZ2vbWfZDwOrJM3N621VbjxwsqS/kzrNPhcRy/IZ5o8lvSmXOw14gtQ2fIukpdHJuywi4llJ9yrdlvky6QqlzXRgUr7sfpyUjLrbmaRtOoTUSb+U9OHqrP9F6jBcRGqP7/SHMyIelHQ1qblxEalZpaP6pwKXSDqedOy0KneOpLGkM+s7SZ23D5Oath5UOrVcRroLaa3jJCLOaxFvcd/dQmquuT+fpb4IHAFsl9e9Gvg78OW8eFeOmS/mbV1JalN/ruRyxZjnSnqI9JlaSGoy64qLgMvzsTkH+E2J+tfa5hblBgM3StqUtK++lqcfD1yQ19eflIgmkfo3fyJpAunqrkzTW1N+1tB6kvRiRAzq7Tis83JSey1fku8JXBQRu/RyWNaEpEER8WIePpV0R90JvRzWBqNWVwRmDUYB10jaiNRJenQvx2OtfULSv5G+sxaR7gyybuIrgjc4SReQ7hsu+u+IuLw34tnQ5f6jO5vM+khENN6x9Iarv04kfRxo/OeNT0bEQc3Kv9HqfyNxIjAzq7m63TVkZmYNnAjMzGrOicDMrOacCMzMau7/A5iEKOSXYVtZAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.bar(names, values)\n",
    "plt.ylabel('accuracy_score')\n",
    "plt.title('Plot the 3 accuracy values')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "4301d16e-0986-458a-8e69-e6c4465f778e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Figure size 432x288 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.savefig('plot_3_accuracy_values.png')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
