{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "21f88c69-0f70-46e5-9bf0-d02bc7edd748",
   "metadata": {},
   "outputs": [],
   "source": [
    "### 2021-10-14 File Created ###\n",
    "### Keagan G Moo            ###\n",
    "### BIOINF 575              ###\n",
    "### Group Project #3        ###\n",
    "\n",
    "# This notebook will use the GEOParse module to load in data from the GEO accession database. For this test\n",
    "# we will use GSE10245 a database on Non Small Cell Lung Cancer cells. The output should be formatted for general\n",
    "# use if it is not already in such a form. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4154a4cb-11e1-4c8f-b5ae-38f3a2c1f727",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: GEOparse in c:\\users\\kgmoo\\anaconda3\\lib\\site-packages (2.0.3)\n",
      "Requirement already satisfied: pandas>=0.17 in c:\\users\\kgmoo\\anaconda3\\lib\\site-packages (from GEOparse) (1.2.4)\n",
      "Requirement already satisfied: tqdm>=4.31.1 in c:\\users\\kgmoo\\anaconda3\\lib\\site-packages (from GEOparse) (4.59.0)\n",
      "Requirement already satisfied: numpy>=1.7 in c:\\users\\kgmoo\\anaconda3\\lib\\site-packages (from GEOparse) (1.20.1)\n",
      "Requirement already satisfied: requests>=2.21.0 in c:\\users\\kgmoo\\anaconda3\\lib\\site-packages (from GEOparse) (2.25.1)\n",
      "Requirement already satisfied: python-dateutil>=2.7.3 in c:\\users\\kgmoo\\anaconda3\\lib\\site-packages (from pandas>=0.17->GEOparse) (2.8.1)\n",
      "Requirement already satisfied: pytz>=2017.3 in c:\\users\\kgmoo\\anaconda3\\lib\\site-packages (from pandas>=0.17->GEOparse) (2021.1)\n",
      "Requirement already satisfied: six>=1.5 in c:\\users\\kgmoo\\anaconda3\\lib\\site-packages (from python-dateutil>=2.7.3->pandas>=0.17->GEOparse) (1.15.0)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in c:\\users\\kgmoo\\anaconda3\\lib\\site-packages (from requests>=2.21.0->GEOparse) (1.26.4)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in c:\\users\\kgmoo\\anaconda3\\lib\\site-packages (from requests>=2.21.0->GEOparse) (2020.12.5)\n",
      "Requirement already satisfied: idna<3,>=2.5 in c:\\users\\kgmoo\\anaconda3\\lib\\site-packages (from requests>=2.21.0->GEOparse) (2.10)\n",
      "Requirement already satisfied: chardet<5,>=3.0.2 in c:\\users\\kgmoo\\anaconda3\\lib\\site-packages (from requests>=2.21.0->GEOparse) (4.0.0)\n"
     ]
    }
   ],
   "source": [
    "# installation method lifted from https://jakevdp.github.io/blog/2017/12/05/installing-python-packages-from-jupyter/\n",
    "# to ensure that this works on different machines. If you are having an issue, try using conda instead of pip.\n",
    "# If you still have issues run \"!conda env list\" in an empty cell and ensure you are using the right python\n",
    "\n",
    "# sys allows you to access system commands of which we need,\n",
    "import sys\n",
    "# sys.executable ensures that you are installing a package to the python instance being used by Jupyter\n",
    "# -m ensures that the you are using the correct instance of pip to perform the installation\n",
    "!{sys.executable} -m pip install GEOparse\n",
    "import GEOparse\n",
    "import pandas as pd\n",
    "\n",
    "slowpoke = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0167e0c8-e809-4a11-bbb3-025c1f2873d1",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "14-Oct-2021 11:09:42 DEBUG utils - Directory ./ already exists. Skipping.\n",
      "14-Oct-2021 11:09:42 INFO GEOparse - File already exist: using local version.\n",
      "14-Oct-2021 11:09:42 INFO GEOparse - Parsing ./GSE10245_family.soft.gz: \n",
      "14-Oct-2021 11:09:42 DEBUG GEOparse - DATABASE: GeoMiame\n",
      "14-Oct-2021 11:09:42 DEBUG GEOparse - SERIES: GSE10245\n",
      "14-Oct-2021 11:09:42 DEBUG GEOparse - PLATFORM: GPL570\n",
      "14-Oct-2021 11:09:44 DEBUG GEOparse - SAMPLE: GSM258551\n",
      "14-Oct-2021 11:09:44 DEBUG GEOparse - SAMPLE: GSM258552\n",
      "14-Oct-2021 11:09:44 DEBUG GEOparse - SAMPLE: GSM258553\n",
      "14-Oct-2021 11:09:44 DEBUG GEOparse - SAMPLE: GSM258554\n",
      "14-Oct-2021 11:09:44 DEBUG GEOparse - SAMPLE: GSM258555\n",
      "14-Oct-2021 11:09:44 DEBUG GEOparse - SAMPLE: GSM258556\n",
      "14-Oct-2021 11:09:45 DEBUG GEOparse - SAMPLE: GSM258557\n",
      "14-Oct-2021 11:09:45 DEBUG GEOparse - SAMPLE: GSM258558\n",
      "14-Oct-2021 11:09:45 DEBUG GEOparse - SAMPLE: GSM258559\n",
      "14-Oct-2021 11:09:45 DEBUG GEOparse - SAMPLE: GSM258560\n",
      "14-Oct-2021 11:09:45 DEBUG GEOparse - SAMPLE: GSM258561\n",
      "14-Oct-2021 11:09:45 DEBUG GEOparse - SAMPLE: GSM258562\n",
      "14-Oct-2021 11:09:45 DEBUG GEOparse - SAMPLE: GSM258563\n",
      "14-Oct-2021 11:09:45 DEBUG GEOparse - SAMPLE: GSM258564\n",
      "14-Oct-2021 11:09:45 DEBUG GEOparse - SAMPLE: GSM258565\n",
      "14-Oct-2021 11:09:46 DEBUG GEOparse - SAMPLE: GSM258566\n",
      "14-Oct-2021 11:09:46 DEBUG GEOparse - SAMPLE: GSM258567\n",
      "14-Oct-2021 11:09:46 DEBUG GEOparse - SAMPLE: GSM258568\n",
      "14-Oct-2021 11:09:46 DEBUG GEOparse - SAMPLE: GSM258569\n",
      "14-Oct-2021 11:09:46 DEBUG GEOparse - SAMPLE: GSM258570\n",
      "14-Oct-2021 11:09:46 DEBUG GEOparse - SAMPLE: GSM258571\n",
      "14-Oct-2021 11:09:46 DEBUG GEOparse - SAMPLE: GSM258572\n",
      "14-Oct-2021 11:09:46 DEBUG GEOparse - SAMPLE: GSM258573\n",
      "14-Oct-2021 11:09:46 DEBUG GEOparse - SAMPLE: GSM258574\n",
      "14-Oct-2021 11:09:47 DEBUG GEOparse - SAMPLE: GSM258575\n",
      "14-Oct-2021 11:09:47 DEBUG GEOparse - SAMPLE: GSM258576\n",
      "14-Oct-2021 11:09:47 DEBUG GEOparse - SAMPLE: GSM258577\n",
      "14-Oct-2021 11:09:47 DEBUG GEOparse - SAMPLE: GSM258578\n",
      "14-Oct-2021 11:09:47 DEBUG GEOparse - SAMPLE: GSM258579\n",
      "14-Oct-2021 11:09:47 DEBUG GEOparse - SAMPLE: GSM258580\n",
      "14-Oct-2021 11:09:47 DEBUG GEOparse - SAMPLE: GSM258581\n",
      "14-Oct-2021 11:09:47 DEBUG GEOparse - SAMPLE: GSM258582\n",
      "14-Oct-2021 11:09:47 DEBUG GEOparse - SAMPLE: GSM258583\n",
      "14-Oct-2021 11:09:47 DEBUG GEOparse - SAMPLE: GSM258584\n",
      "14-Oct-2021 11:09:48 DEBUG GEOparse - SAMPLE: GSM258585\n",
      "14-Oct-2021 11:09:48 DEBUG GEOparse - SAMPLE: GSM258586\n",
      "14-Oct-2021 11:09:48 DEBUG GEOparse - SAMPLE: GSM258587\n",
      "14-Oct-2021 11:09:48 DEBUG GEOparse - SAMPLE: GSM258588\n",
      "14-Oct-2021 11:09:48 DEBUG GEOparse - SAMPLE: GSM258589\n",
      "14-Oct-2021 11:09:48 DEBUG GEOparse - SAMPLE: GSM258590\n",
      "14-Oct-2021 11:09:48 DEBUG GEOparse - SAMPLE: GSM258591\n",
      "14-Oct-2021 11:09:48 DEBUG GEOparse - SAMPLE: GSM258592\n",
      "14-Oct-2021 11:09:48 DEBUG GEOparse - SAMPLE: GSM258593\n",
      "14-Oct-2021 11:09:49 DEBUG GEOparse - SAMPLE: GSM258594\n",
      "14-Oct-2021 11:09:49 DEBUG GEOparse - SAMPLE: GSM258595\n",
      "14-Oct-2021 11:09:49 DEBUG GEOparse - SAMPLE: GSM258596\n",
      "14-Oct-2021 11:09:49 DEBUG GEOparse - SAMPLE: GSM258597\n",
      "14-Oct-2021 11:09:49 DEBUG GEOparse - SAMPLE: GSM258598\n",
      "14-Oct-2021 11:09:49 DEBUG GEOparse - SAMPLE: GSM258599\n",
      "14-Oct-2021 11:09:49 DEBUG GEOparse - SAMPLE: GSM258600\n",
      "14-Oct-2021 11:09:49 DEBUG GEOparse - SAMPLE: GSM258601\n",
      "14-Oct-2021 11:09:49 DEBUG GEOparse - SAMPLE: GSM258602\n",
      "14-Oct-2021 11:09:49 DEBUG GEOparse - SAMPLE: GSM258603\n",
      "14-Oct-2021 11:09:50 DEBUG GEOparse - SAMPLE: GSM258604\n",
      "14-Oct-2021 11:09:50 DEBUG GEOparse - SAMPLE: GSM258605\n",
      "14-Oct-2021 11:09:50 DEBUG GEOparse - SAMPLE: GSM258606\n",
      "14-Oct-2021 11:09:50 DEBUG GEOparse - SAMPLE: GSM258607\n",
      "14-Oct-2021 11:09:50 DEBUG GEOparse - SAMPLE: GSM258608\n"
     ]
    }
   ],
   "source": [
    "# Use the lovely GEO API to get the default version of the data at that address, in this case it is the\n",
    "# soft zipped file which is what we want\n",
    "rawDataGSE10245 = GEOparse.get_GEO(geo=\"GSE10245\", destdir=\"./\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "e3042709-e3c2-449b-91cf-d9ca4ed66a44",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'GEOparse.GEOTypes.GSE'>\n",
      "Non-small cell lung cancer (NSCLC) can be classified into the major subtypes adenocarcinoma (AC) and squamous cell carcinoma (SCC) subtypes. Although explicit molecular, histological and clinical characteristics have been reported for both subtypes, no specific therapy exists so far. However, the characterization of suitable molecular targets holds great promises to develop novel therapies in NSCLC. In the present study, global gene expression profiling of 58 human high grade NSCLC specimens revealed large transcriptomic differences between AC and SCC subtypes: More than 1.700 genes were found to be differentially expressed.\n",
      "\n",
      "Sample GSM258551\n",
      "('title', ['NSCLC_AC_10'])\n",
      "('geo_accession', ['GSM258551'])\n",
      "('status', ['Public on Oct 01 2009'])\n",
      "('submission_date', ['Jan 23 2008'])\n",
      "('last_update_date', ['Aug 28 2018'])\n",
      "('type', ['RNA'])\n",
      "('channel_count', ['1'])\n",
      "('source_name_ch1', ['human non-small cell lung cancer tumor tissue'])\n",
      "('organism_ch1', ['Homo sapiens'])\n",
      "('taxid_ch1', ['9606'])\n",
      "('characteristics_ch1', ['disease state: adenocarcinoma'])\n",
      "('treatment_protocol_ch1', ['tumor tissue samples from surgically treated patients were collected intraoperatively, snap-frozen and stored at -80 C. All samples underwent a careful analysis of the histopathology, estimation of tumor and stromal content and subsequent macro-dissection of the tumor regions before enrollment into the microarray experiments.'])\n",
      "('molecule_ch1', ['total RNA'])\n",
      "('extract_protocol_ch1', ['total RNA was extracted using the RNeasy Kit (Qiagen, Hilden, Germany) according to the manufacturersÃ† instructions.'])\n",
      "('label_ch1', ['biotin'])\n",
      "('label_protocol_ch1', ['Biotinylated cRNA were prepared according to the standard Affymetrix protocol from 2 ug total RNA (Affymetrix ROC technical manual 701021 Rev.5).'])\n",
      "('hyb_protocol', ['Following fragmentation, 10 ug of cRNA were hybridized for 16 hr at 45C on Affymetrix GeneChipÂ« U133 Plus 2.0 in the Affymetrix GeneChipÂ« Hybridization Oven 640. GeneChips were washed and stained in the Affymetrix GeneChipÂ« Fluidics Station 450.'])\n",
      "('scan_protocol', ['Hybridized microarrays were scanned with the Affymetrix GeneChipÂ« Scanner 3000.'])\n",
      "('description', ['Gene expression data from human non-small cell lung cancer tumor tissue 10.'])\n",
      "('data_processing', ['Bioconductor [Gentleman et al. Genome Biology 2004] and R language [Ihaka and Gentleman. Journal of Computational and Graphical Statistics 1996] were used for pre-processing of the 58 hybridizations. The microarray data was normalized by the gcRMA method [Wu et al. Journal of the American Statistical Association 2004].'])\n",
      "('platform_id', ['GPL570'])\n",
      "('contact_name', ['Ruprecht,,Kuner'])\n",
      "('contact_laboratory', ['Unit Cancer Genome Research'])\n",
      "('contact_department', ['Molecular Genetics'])\n",
      "('contact_institute', ['German Cancer Research Center and National Center of Tumor Diseases'])\n",
      "('contact_address', ['Im Neuenheimer Feld 460'])\n",
      "('contact_city', ['Heidelberg'])\n",
      "('contact_zip/postal_code', ['69120'])\n",
      "('contact_country', ['Germany'])\n",
      "('supplementary_file', ['ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM258nnn/GSM258551/suppl/GSM258551.CEL.gz'])\n",
      "('relation', ['Reanalyzed by: GSE60486', 'Reanalyzed by: GSE119087'])\n",
      "('series_id', ['GSE10245'])\n",
      "('data_row_count', ['54675'])\n",
      "\n",
      "Platform GPL570\n",
      "('title', ['[HG-U133_Plus_2] Affymetrix Human Genome U133 Plus 2.0 Array'])\n",
      "('geo_accession', ['GPL570'])\n",
      "('status', ['Public on Nov 07 2003'])\n",
      "('submission_date', ['Nov 07 2003'])\n",
      "('last_update_date', ['Dec 14 2020'])\n",
      "('technology', ['in situ oligonucleotide'])\n",
      "('distribution', ['commercial'])\n",
      "('organism', ['Homo sapiens'])\n",
      "('taxid', ['9606'])\n",
      "('manufacturer', ['Affymetrix'])\n",
      "('manufacture_protocol', [\"see manufacturer's web site\", '', 'Complete coverage of the Human Genome U133 Set plus 6,500 additional genes for analysis of over 47,000 transcripts', 'All probe sets represented on the GeneChip Human Genome U133 Set are identically replicated on the GeneChip Human Genome U133 Plus 2.0 Array. The sequences from which these probe sets were derived were selected from GenBankÂ®, dbEST, and RefSeq. The sequence clusters were created from the UniGene database (Build 133, April 20, 2001) and then refined by analysis and comparison with a number of other publicly available databases, including the Washington University EST trace repository and the University of California, Santa Cruz Golden-Path human genome database (April 2001 release).', 'In addition, there are 9,921 new probe sets representing approximately 6,500 new genes. These gene sequences were selected from GenBank, dbEST, and RefSeq. Sequence clusters were created from the UniGene database (Build 159, January 25, 2003) and refined by analysis and comparison with a number of other publicly available databases, including the Washington University EST trace repository and the NCBI human genome assembly (Build 31).', ''])\n",
      "('description', ['Affymetrix submissions are typically submitted to GEO using the GEOarchive method described at http://www.ncbi.nlm.nih.gov/projects/geo/info/geo_affy.html', '', 'June 03, 2009: annotation table updated with netaffx build 28', 'June 06, 2012: annotation table updated with netaffx build 32', 'June 23, 2016: annotation table updated with netaffx build 35'])\n",
      "('web_link', ['http://www.affymetrix.com/support/technical/byproduct.affx?product=hg-u133-plus', 'http://www.affymetrix.com/analysis/index.affx'])\n",
      "('contact_name', [',,Affymetrix, Inc.'])\n",
      "('contact_email', ['geo@ncbi.nlm.nih.gov, support@affymetrix.com'])\n",
      "('contact_phone', ['888-362-2447'])\n",
      "('contact_institute', ['Affymetrix, Inc.'])\n",
      "('contact_address', [''])\n",
      "('contact_city', ['Santa Clara'])\n",
      "('contact_state', ['CA'])\n",
      "('contact_zip/postal_code', ['95051'])\n",
      "('contact_country', ['USA'])\n",
      "('contact_web_link', ['http://www.affymetrix.com/index.affx'])\n",
      "('relation', ['Alternative to: GPL4454 (Alternative CDF)', 'Alternative to: GPL4866 (Alternative CDF)', 'Alternative to: GPL5760 (Alternative CDF)', 'Alternative to: GPL6671 (Alternative CDF)', 'Alternative to: GPL6791 (Alternative CDF)', 'Alternative to: GPL6879 (Alternative CDF)', 'Alternative to: GPL7567 (Alternative CDF)', 'Alternative to: GPL8019 (Alternative CDF)', 'Alternative to: GPL8542 (Alternative CDF)', 'Alternative to: GPL8715 (Alternative CDF)', 'Alternative to: GPL8712 (Alternative CDF)', 'Alternative to: GPL9102 (Probe Level Version)', 'Alternative to: GPL9099 (Alternative CDF)', 'Alternative to: GPL9101 (Alternative CDF)', 'Alternative to: GPL9324 (Alternative CDF)', 'Alternative to: GPL9486 (Alternative CDF)', 'Alternative to: GPL9987 (Alternative CDF)', 'Alternative to: GPL10175 (Alternative CDF)', 'Alternative to: GPL10335 (Alternative CDF)', 'Alternative to: GPL10371 (Alternative CDF)', 'Alternative to: GPL10526 (Alternative CDF)', 'Alternative to: GPL10881 (Alternative CDF)', 'Alternative to: GPL10925 (Alternative CDF)', 'Alternative to: GPL11084 (Alternative CDF)', 'Alternative to: GPL11433 (Alternative CDF)', 'Alternative to: GPL11670 (Alternative CDF)', 'Alternative to: GPL13232 (Alternative CDF [HGU133Plus2_Hs_REFSEQ, Brainarray version 13.0.0])', 'Alternative to: GPL13303 (Alternative CDF.)', 'Alternative to: GPL13916 ([Brainarray ENSG Version 14.1])', 'Alternative to: GPL14837 (Alternative CDF)', 'Alternative to: GPL14877 ([Alternative CDF])', 'Alternative to: GPL15394 (Alternative CDF [HGU133Plus2_Hs_REFSEQ, Brainarray version 14.0])', 'Alternative to: GPL6732 (Alternative CDF [HGU133Plus2_Hs_REFSEQ, Brainarray version 7])', 'Alternative to: GPL15445 (Alternative CDF)', 'Alternative to: GPL16066 (Alternative CDF)', 'Alternative to: GPL16100 (Alternative CDF [GATExplorer_Ensembl v57])', 'Alternative to: GPL16268 (Alternative CDF [Brainarray ENSG Version 16])', 'Alternative to: GPL16273 (Alternative CDF [Brainarray ENSG Version 15])', 'Alternative to: GPL16311 (Alternative CDF [HGU133Plus2_Hs_ENTREZG Brainarray version 14.0.0])', 'Alternative to: GPL16356 (Alternative CDF [Brainarray HGU133Plus2_Hs_ENTREZG 15.0.0])', 'Alternative to: GPL16372 (Alternative CDF [Brainarray HGU133Plus2_Hs_UG v.13])', 'Alternative to: GPL17392 (Alternative CDF [HGU133Plus2_Hs_ENSG, Brainarray 17.1])', 'Alternative to: GPL17394 (Alternative CDF [HGU133Plus2_Hs_ENSG, Brainarray 13.0])', 'Alternative to: GPL17810 (Alternative CDF [HGU133Plus2_Hs_ENTREZG, Brainarray 16])', 'Alternative to: GPL17811 (Alternative CDF [HGU133Plus2_Hs_ENTREZG, Brainarray 17])', 'Alternative to: GPL17996 (Alternative CDF:HGU133Plus2_Hs_ENTREZG.cdf version 15.1.0)', 'Alternative to: GPL18121 (custom CDF)', 'Alternative to: GPL18756 (Alternative CDF [CDF: HGU133Plus2_Hs_ENTREZG_14.1.0])', 'Alternative to: GPL18850 (Alternative CDF)', 'Alternative to: GPL18478 (Alternative CDF [Hs_ENTREZG_V13])', 'Alternative to: GPL19109 (Alternative CDF [Brainarray HGU133Plus2_Hs_ENTREZG_v18])', 'Alternative to: GPL19883 (Alternative CDF [Broad GenePattern])', 'Alternative to: GPL19918 (Gene symbol version, 10K)', 'Alternative to: GPL20182 (Gene Symbol Version)', 'Alternative to: GPL22321 (Alternative CDF [HGU133Plus2_Hs_ENTREZG, Brainarray 20])', 'Alternative to: GPL22945 (alternative)', 'Alternative to: GPL23270 ((Alternative CDF [HGU133Plus2_Hs_ENTREZG, Brainarray 21])', 'Alternative to: GPL23432 (alternative CDF [HGU133Plus2_Hs_ENSG, Brainarray version 18.0.0])', 'Alternative to: GPL25740 (alternative)', 'Alternative to: GPL28718 (Alternative CDF [HGU133Plus2_Hs_ENTREZG, Brainarray 23])', 'Alternative to: GPL29499 (Alternative CDF)'])\n",
      "('data_row_count', ['54675'])\n"
     ]
    }
   ],
   "source": [
    "# What we have here is a complex bespoke object with some finnicky properties, let's unpack that\n",
    "if slowpoke:\n",
    "    print(type(rawDataGSE10245))\n",
    "    print(rawDataGSE10245.metadata['summary'][0])\n",
    "    print(\"\\nSample GSM258551\")\n",
    "    for metaIndex in rawDataGSE10245.gsms.get(\"GSM258551\").metadata.items():\n",
    "        print(str(metaIndex))\n",
    "    print(\"\\nPlatform GPL570\")\n",
    "    for metaIndex in rawDataGSE10245.gpls.get('GPL570').metadata.items():\n",
    "        print(str(metaIndex))\n",
    "        \n",
    "# Important details include\n",
    "# 1 Platform, meaning that this was all done on the same machine and is therefore comparable\n",
    "# Homo Sapiens as species, informing what references to use for our own analysis\n",
    "# Year of original analysis, informing what references to use for comparative analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "123dda2a-c765-4fc4-bd08-286822b6612b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID_REF</th>\n",
       "      <th>VALUE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1007_s_at</td>\n",
       "      <td>9.129905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1053_at</td>\n",
       "      <td>8.034022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>117_at</td>\n",
       "      <td>3.564520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>121_at</td>\n",
       "      <td>4.746490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1255_g_at</td>\n",
       "      <td>2.320698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54670</th>\n",
       "      <td>AFFX-r2-Ec-bioC-5_at</td>\n",
       "      <td>10.730983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54671</th>\n",
       "      <td>AFFX-r2-Ec-bioD-3_at</td>\n",
       "      <td>13.599488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54672</th>\n",
       "      <td>AFFX-r2-Ec-bioD-5_at</td>\n",
       "      <td>13.031726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54673</th>\n",
       "      <td>AFFX-r2-P1-cre-3_at</td>\n",
       "      <td>15.028729</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54674</th>\n",
       "      <td>AFFX-r2-P1-cre-5_at</td>\n",
       "      <td>14.586347</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54675 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     ID_REF      VALUE\n",
       "0                 1007_s_at   9.129905\n",
       "1                   1053_at   8.034022\n",
       "2                    117_at   3.564520\n",
       "3                    121_at   4.746490\n",
       "4                 1255_g_at   2.320698\n",
       "...                     ...        ...\n",
       "54670  AFFX-r2-Ec-bioC-5_at  10.730983\n",
       "54671  AFFX-r2-Ec-bioD-3_at  13.599488\n",
       "54672  AFFX-r2-Ec-bioD-5_at  13.031726\n",
       "54673   AFFX-r2-P1-cre-3_at  15.028729\n",
       "54674   AFFX-r2-P1-cre-5_at  14.586347\n",
       "\n",
       "[54675 rows x 2 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The actual data we want to analyze however is stored in each sample seperately\n",
    "rawDataGSE10245.gsms.get('GSM258551').table"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
